// @aws-sdk/client-cost-and-usage-report-service/dist-es/models/models_0.js
import { CostAndUsageReportServiceServiceException as __BaseException } from "./CostAndUsageReportServiceServiceException";
export const AdditionalArtifact = {
    ATHENA: "ATHENA",
    QUICKSIGHT: "QUICKSIGHT",
    REDSHIFT: "REDSHIFT",
};
export class InternalErrorException extends __BaseException {
    name = "InternalErrorException";
    $fault = "server";
    Message;
    constructor(opts) {
        super({
            name: "InternalErrorException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalErrorException.prototype);
        this.Message = opts.Message;
    }
}
export class ValidationException extends __BaseException {
    name = "ValidationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ValidationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ValidationException.prototype);
        this.Message = opts.Message;
    }
}
export const SchemaElement = {
    MANUAL_DISCOUNT_COMPATIBILITY: "MANUAL_DISCOUNT_COMPATIBILITY",
    RESOURCES: "RESOURCES",
    SPLIT_COST_ALLOCATION_DATA: "SPLIT_COST_ALLOCATION_DATA",
};
export const CompressionFormat = {
    GZIP: "GZIP",
    Parquet: "Parquet",
    ZIP: "ZIP",
};
export const ReportFormat = {
    CSV: "textORcsv",
    Parquet: "Parquet",
};
export const LastStatus = {
    ERROR_NO_BUCKET: "ERROR_NO_BUCKET",
    ERROR_PERMISSIONS: "ERROR_PERMISSIONS",
    SUCCESS: "SUCCESS",
};
export const ReportVersioning = {
    CREATE_NEW_REPORT: "CREATE_NEW_REPORT",
    OVERWRITE_REPORT: "OVERWRITE_REPORT",
};
export const AWSRegion = {
    BAHRAIN: "me-south-1",
    BEIJING: "cn-north-1",
    CANADA_CENTRAL: "ca-central-1",
    CAPE_TOWN: "af-south-1",
    FRANKFURT: "eu-central-1",
    HONG_KONG: "ap-east-1",
    HYDERABAD: "ap-south-2",
    IRELAND: "eu-west-1",
    JAKARTA: "ap-southeast-3",
    LONDON: "eu-west-2",
    MILANO: "eu-south-1",
    MUMBAI: "ap-south-1",
    NINGXIA: "cn-northwest-1",
    NORTHERN_CALIFORNIA: "us-west-1",
    OHIO: "us-east-2",
    OREGON: "us-west-2",
    OSAKA: "ap-northeast-3",
    PARIS: "eu-west-3",
    SAO_PAULO: "sa-east-1",
    SEOUL: "ap-northeast-2",
    SINGAPORE: "ap-southeast-1",
    SPAIN: "eu-south-2",
    STOCKHOLM: "eu-north-1",
    SYDNEY: "ap-southeast-2",
    TOKYO: "ap-northeast-1",
    UAE: "me-central-1",
    US_STANDARD: "us-east-1",
    ZURICH: "eu-central-2",
};
export const TimeUnit = {
    DAILY: "DAILY",
    HOURLY: "HOURLY",
    MONTHLY: "MONTHLY",
};
export class ResourceNotFoundException extends __BaseException {
    name = "ResourceNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
export class DuplicateReportNameException extends __BaseException {
    name = "DuplicateReportNameException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "DuplicateReportNameException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DuplicateReportNameException.prototype);
        this.Message = opts.Message;
    }
}
export class ReportLimitReachedException extends __BaseException {
    name = "ReportLimitReachedException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ReportLimitReachedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ReportLimitReachedException.prototype);
        this.Message = opts.Message;
    }
}
