// @aws-sdk/client-organizations/dist-cjs/index.js
'use strict';

var middlewareHostHeader = require('@aws-sdk/middleware-host-header');
var middlewareLogger = require('@aws-sdk/middleware-logger');
var middlewareRecursionDetection = require('@aws-sdk/middleware-recursion-detection');
var middlewareUserAgent = require('@aws-sdk/middleware-user-agent');
var configResolver = require('@smithy/config-resolver');
var core = require('@smithy/core');
var middlewareContentLength = require('@smithy/middleware-content-length');
var middlewareEndpoint = require('@smithy/middleware-endpoint');
var middlewareRetry = require('@smithy/middleware-retry');
var smithyClient = require('@smithy/smithy-client');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var regionConfigResolver = require('@aws-sdk/region-config-resolver');
var protocolHttp = require('@smithy/protocol-http');
var middlewareSerde = require('@smithy/middleware-serde');
var core$1 = require('@aws-sdk/core');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "organizations",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(regionConfigResolver.getAwsRegionExtensionConfiguration(runtimeConfig), smithyClient.getDefaultExtensionConfiguration(runtimeConfig), protocolHttp.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, regionConfigResolver.resolveAwsRegionExtensionConfiguration(extensionConfiguration), smithyClient.resolveDefaultRuntimeConfig(extensionConfiguration), protocolHttp.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class OrganizationsClient extends smithyClient.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = middlewareUserAgent.resolveUserAgentConfig(_config_1);
        const _config_3 = middlewareRetry.resolveRetryConfig(_config_2);
        const _config_4 = configResolver.resolveRegionConfig(_config_3);
        const _config_5 = middlewareHostHeader.resolveHostHeaderConfig(_config_4);
        const _config_6 = middlewareEndpoint.resolveEndpointConfig(_config_5);
        const _config_7 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(middlewareUserAgent.getUserAgentPlugin(this.config));
        this.middlewareStack.use(middlewareRetry.getRetryPlugin(this.config));
        this.middlewareStack.use(middlewareContentLength.getContentLengthPlugin(this.config));
        this.middlewareStack.use(middlewareHostHeader.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(middlewareLogger.getLoggerPlugin(this.config));
        this.middlewareStack.use(middlewareRecursionDetection.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultOrganizationsHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class OrganizationsServiceException extends smithyClient.ServiceException {
    constructor(options) {
        super(options);
        Object.setPrototypeOf(this, OrganizationsServiceException.prototype);
    }
}

const ActionType = {
    ADD_ORGANIZATIONS_SERVICE_LINKED_ROLE: "ADD_ORGANIZATIONS_SERVICE_LINKED_ROLE",
    APPROVE_ALL_FEATURES: "APPROVE_ALL_FEATURES",
    ENABLE_ALL_FEATURES: "ENABLE_ALL_FEATURES",
    INVITE_ACCOUNT_TO_ORGANIZATION: "INVITE",
};
const HandshakePartyType = {
    ACCOUNT: "ACCOUNT",
    EMAIL: "EMAIL",
    ORGANIZATION: "ORGANIZATION",
};
const HandshakeResourceType = {
    ACCOUNT: "ACCOUNT",
    EMAIL: "EMAIL",
    MASTER_EMAIL: "MASTER_EMAIL",
    MASTER_NAME: "MASTER_NAME",
    NOTES: "NOTES",
    ORGANIZATION: "ORGANIZATION",
    ORGANIZATION_FEATURE_SET: "ORGANIZATION_FEATURE_SET",
    PARENT_HANDSHAKE: "PARENT_HANDSHAKE",
};
const HandshakeState = {
    ACCEPTED: "ACCEPTED",
    CANCELED: "CANCELED",
    DECLINED: "DECLINED",
    EXPIRED: "EXPIRED",
    OPEN: "OPEN",
    REQUESTED: "REQUESTED",
};
class AccessDeniedException extends OrganizationsServiceException {
    name = "AccessDeniedException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
        this.Message = opts.Message;
    }
}
const AccessDeniedForDependencyExceptionReason = {
    ACCESS_DENIED_DURING_CREATE_SERVICE_LINKED_ROLE: "ACCESS_DENIED_DURING_CREATE_SERVICE_LINKED_ROLE",
};
class AccessDeniedForDependencyException extends OrganizationsServiceException {
    name = "AccessDeniedForDependencyException";
    $fault = "client";
    Message;
    Reason;
    constructor(opts) {
        super({
            name: "AccessDeniedForDependencyException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedForDependencyException.prototype);
        this.Message = opts.Message;
        this.Reason = opts.Reason;
    }
}
class AWSOrganizationsNotInUseException extends OrganizationsServiceException {
    name = "AWSOrganizationsNotInUseException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AWSOrganizationsNotInUseException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AWSOrganizationsNotInUseException.prototype);
        this.Message = opts.Message;
    }
}
class ConcurrentModificationException extends OrganizationsServiceException {
    name = "ConcurrentModificationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ConcurrentModificationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConcurrentModificationException.prototype);
        this.Message = opts.Message;
    }
}
class HandshakeAlreadyInStateException extends OrganizationsServiceException {
    name = "HandshakeAlreadyInStateException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "HandshakeAlreadyInStateException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, HandshakeAlreadyInStateException.prototype);
        this.Message = opts.Message;
    }
}
const HandshakeConstraintViolationExceptionReason = {
    ACCOUNT_NUMBER_LIMIT_EXCEEDED: "ACCOUNT_NUMBER_LIMIT_EXCEEDED",
    ALREADY_IN_AN_ORGANIZATION: "ALREADY_IN_AN_ORGANIZATION",
    HANDSHAKE_RATE_LIMIT_EXCEEDED: "HANDSHAKE_RATE_LIMIT_EXCEEDED",
    INVITE_DISABLED_DURING_ENABLE_ALL_FEATURES: "INVITE_DISABLED_DURING_ENABLE_ALL_FEATURES",
    MANAGEMENT_ACCOUNT_EMAIL_NOT_VERIFIED: "MANAGEMENT_ACCOUNT_EMAIL_NOT_VERIFIED",
    ORGANIZATION_ALREADY_HAS_ALL_FEATURES: "ORGANIZATION_ALREADY_HAS_ALL_FEATURES",
    ORGANIZATION_FROM_DIFFERENT_SELLER_OF_RECORD: "ORGANIZATION_FROM_DIFFERENT_SELLER_OF_RECORD",
    ORGANIZATION_IS_ALREADY_PENDING_ALL_FEATURES_MIGRATION: "ORGANIZATION_IS_ALREADY_PENDING_ALL_FEATURES_MIGRATION",
    ORGANIZATION_MEMBERSHIP_CHANGE_RATE_LIMIT_EXCEEDED: "ORGANIZATION_MEMBERSHIP_CHANGE_RATE_LIMIT_EXCEEDED",
    PAYMENT_INSTRUMENT_REQUIRED: "PAYMENT_INSTRUMENT_REQUIRED",
};
class HandshakeConstraintViolationException extends OrganizationsServiceException {
    name = "HandshakeConstraintViolationException";
    $fault = "client";
    Message;
    Reason;
    constructor(opts) {
        super({
            name: "HandshakeConstraintViolationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, HandshakeConstraintViolationException.prototype);
        this.Message = opts.Message;
        this.Reason = opts.Reason;
    }
}
class HandshakeNotFoundException extends OrganizationsServiceException {
    name = "HandshakeNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "HandshakeNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, HandshakeNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class InvalidHandshakeTransitionException extends OrganizationsServiceException {
    name = "InvalidHandshakeTransitionException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "InvalidHandshakeTransitionException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidHandshakeTransitionException.prototype);
        this.Message = opts.Message;
    }
}
const InvalidInputExceptionReason = {
    DUPLICATE_TAG_KEY: "DUPLICATE_TAG_KEY",
    IMMUTABLE_POLICY: "IMMUTABLE_POLICY",
    INPUT_REQUIRED: "INPUT_REQUIRED",
    INVALID_EMAIL_ADDRESS_TARGET: "INVALID_EMAIL_ADDRESS_TARGET",
    INVALID_ENUM: "INVALID_ENUM",
    INVALID_ENUM_POLICY_TYPE: "INVALID_ENUM_POLICY_TYPE",
    INVALID_FULL_NAME_TARGET: "INVALID_FULL_NAME_TARGET",
    INVALID_LIST_MEMBER: "INVALID_LIST_MEMBER",
    INVALID_PAGINATION_TOKEN: "INVALID_NEXT_TOKEN",
    INVALID_PARTY_TYPE_TARGET: "INVALID_PARTY_TYPE_TARGET",
    INVALID_PATTERN: "INVALID_PATTERN",
    INVALID_PATTERN_TARGET_ID: "INVALID_PATTERN_TARGET_ID",
    INVALID_PRINCIPAL: "INVALID_PRINCIPAL",
    INVALID_RESOURCE_POLICY_JSON: "INVALID_RESOURCE_POLICY_JSON",
    INVALID_ROLE_NAME: "INVALID_ROLE_NAME",
    INVALID_SYNTAX_ORGANIZATION: "INVALID_SYNTAX_ORGANIZATION_ARN",
    INVALID_SYNTAX_POLICY: "INVALID_SYNTAX_POLICY_ID",
    INVALID_SYSTEM_TAGS_PARAMETER: "INVALID_SYSTEM_TAGS_PARAMETER",
    MAX_FILTER_LIMIT_EXCEEDED: "MAX_LIMIT_EXCEEDED_FILTER",
    MAX_LENGTH_EXCEEDED: "MAX_LENGTH_EXCEEDED",
    MAX_VALUE_EXCEEDED: "MAX_VALUE_EXCEEDED",
    MIN_LENGTH_EXCEEDED: "MIN_LENGTH_EXCEEDED",
    MIN_VALUE_EXCEEDED: "MIN_VALUE_EXCEEDED",
    MOVING_ACCOUNT_BETWEEN_DIFFERENT_ROOTS: "MOVING_ACCOUNT_BETWEEN_DIFFERENT_ROOTS",
    NON_DETACHABLE_POLICY: "NON_DETACHABLE_POLICY",
    TARGET_NOT_SUPPORTED: "TARGET_NOT_SUPPORTED",
    UNRECOGNIZED_SERVICE_PRINCIPAL: "UNRECOGNIZED_SERVICE_PRINCIPAL",
    UNSUPPORTED_ACTION_IN_RESOURCE_POLICY: "UNSUPPORTED_ACTION_IN_RESOURCE_POLICY",
    UNSUPPORTED_POLICY_TYPE_IN_RESOURCE_POLICY: "UNSUPPORTED_POLICY_TYPE_IN_RESOURCE_POLICY",
    UNSUPPORTED_RESOURCE_IN_RESOURCE_POLICY: "UNSUPPORTED_RESOURCE_IN_RESOURCE_POLICY",
};
class InvalidInputException extends OrganizationsServiceException {
    name = "InvalidInputException";
    $fault = "client";
    Message;
    Reason;
    constructor(opts) {
        super({
            name: "InvalidInputException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidInputException.prototype);
        this.Message = opts.Message;
        this.Reason = opts.Reason;
    }
}
class ServiceException extends OrganizationsServiceException {
    name = "ServiceException";
    $fault = "server";
    Message;
    constructor(opts) {
        super({
            name: "ServiceException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceException.prototype);
        this.Message = opts.Message;
    }
}
class TooManyRequestsException extends OrganizationsServiceException {
    name = "TooManyRequestsException";
    $fault = "client";
    Type;
    Message;
    constructor(opts) {
        super({
            name: "TooManyRequestsException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TooManyRequestsException.prototype);
        this.Type = opts.Type;
        this.Message = opts.Message;
    }
}
const AccountJoinedMethod = {
    CREATED: "CREATED",
    INVITED: "INVITED",
};
const AccountState = {
    ACTIVE: "ACTIVE",
    CLOSED: "CLOSED",
    PENDING_ACTIVATION: "PENDING_ACTIVATION",
    PENDING_CLOSURE: "PENDING_CLOSURE",
    SUSPENDED: "SUSPENDED",
};
const AccountStatus = {
    ACTIVE: "ACTIVE",
    PENDING_CLOSURE: "PENDING_CLOSURE",
    SUSPENDED: "SUSPENDED",
};
class AccountAlreadyClosedException extends OrganizationsServiceException {
    name = "AccountAlreadyClosedException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AccountAlreadyClosedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccountAlreadyClosedException.prototype);
        this.Message = opts.Message;
    }
}
class AccountAlreadyRegisteredException extends OrganizationsServiceException {
    name = "AccountAlreadyRegisteredException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AccountAlreadyRegisteredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccountAlreadyRegisteredException.prototype);
        this.Message = opts.Message;
    }
}
class AccountNotFoundException extends OrganizationsServiceException {
    name = "AccountNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AccountNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccountNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class AccountNotRegisteredException extends OrganizationsServiceException {
    name = "AccountNotRegisteredException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AccountNotRegisteredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccountNotRegisteredException.prototype);
        this.Message = opts.Message;
    }
}
class AccountOwnerNotVerifiedException extends OrganizationsServiceException {
    name = "AccountOwnerNotVerifiedException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AccountOwnerNotVerifiedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccountOwnerNotVerifiedException.prototype);
        this.Message = opts.Message;
    }
}
class AlreadyInOrganizationException extends OrganizationsServiceException {
    name = "AlreadyInOrganizationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AlreadyInOrganizationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AlreadyInOrganizationException.prototype);
        this.Message = opts.Message;
    }
}
const ConstraintViolationExceptionReason = {
    ACCOUNT_CANNOT_LEAVE_ORGANIZATION: "ACCOUNT_CANNOT_LEAVE_ORGANIZATION",
    ACCOUNT_CANNOT_LEAVE_WITHOUT_EULA: "ACCOUNT_CANNOT_LEAVE_WITHOUT_EULA",
    ACCOUNT_CANNOT_LEAVE_WITHOUT_PHONE_VERIFICATION: "ACCOUNT_CANNOT_LEAVE_WITHOUT_PHONE_VERIFICATION",
    ACCOUNT_CREATION_NOT_COMPLETE: "ACCOUNT_CREATION_NOT_COMPLETE",
    ACCOUNT_CREATION_RATE_LIMIT_EXCEEDED: "ACCOUNT_CREATION_RATE_LIMIT_EXCEEDED",
    ACCOUNT_NUMBER_LIMIT_EXCEEDED: "ACCOUNT_NUMBER_LIMIT_EXCEEDED",
    ALL_FEATURES_MIGRATION_ORGANIZATION_SIZE_LIMIT_EXCEEDED: "ALL_FEATURES_MIGRATION_ORGANIZATION_SIZE_LIMIT_EXCEEDED",
    CANNOT_CLOSE_MANAGEMENT_ACCOUNT: "CANNOT_CLOSE_MANAGEMENT_ACCOUNT",
    CANNOT_REGISTER_MASTER_AS_DELEGATED_ADMINISTRATOR: "CANNOT_REGISTER_MASTER_AS_DELEGATED_ADMINISTRATOR",
    CANNOT_REGISTER_SUSPENDED_ACCOUNT_AS_DELEGATED_ADMINISTRATOR: "CANNOT_REGISTER_SUSPENDED_ACCOUNT_AS_DELEGATED_ADMINISTRATOR",
    CANNOT_REMOVE_DELEGATED_ADMINISTRATOR_FROM_ORG: "CANNOT_REMOVE_DELEGATED_ADMINISTRATOR_FROM_ORG",
    CLOSE_ACCOUNT_QUOTA_EXCEEDED: "CLOSE_ACCOUNT_QUOTA_EXCEEDED",
    CLOSE_ACCOUNT_REQUESTS_LIMIT_EXCEEDED: "CLOSE_ACCOUNT_REQUESTS_LIMIT_EXCEEDED",
    CREATE_ORGANIZATION_IN_BILLING_MODE_UNSUPPORTED_REGION: "CREATE_ORGANIZATION_IN_BILLING_MODE_UNSUPPORTED_REGION",
    DELEGATED_ADMINISTRATOR_EXISTS_FOR_THIS_SERVICE: "DELEGATED_ADMINISTRATOR_EXISTS_FOR_THIS_SERVICE",
    EMAIL_VERIFICATION_CODE_EXPIRED: "EMAIL_VERIFICATION_CODE_EXPIRED",
    HANDSHAKE_RATE_LIMIT_EXCEEDED: "HANDSHAKE_RATE_LIMIT_EXCEEDED",
    INVALID_PAYMENT_INSTRUMENT: "INVALID_PAYMENT_INSTRUMENT",
    MASTER_ACCOUNT_ADDRESS_DOES_NOT_MATCH_MARKETPLACE: "MASTER_ACCOUNT_ADDRESS_DOES_NOT_MATCH_MARKETPLACE",
    MASTER_ACCOUNT_MISSING_BUSINESS_LICENSE: "MASTER_ACCOUNT_MISSING_BUSINESS_LICENSE",
    MASTER_ACCOUNT_MISSING_CONTACT_INFO: "MASTER_ACCOUNT_MISSING_CONTACT_INFO",
    MASTER_ACCOUNT_NOT_GOVCLOUD_ENABLED: "MASTER_ACCOUNT_NOT_GOVCLOUD_ENABLED",
    MASTER_ACCOUNT_PAYMENT_INSTRUMENT_REQUIRED: "MASTER_ACCOUNT_PAYMENT_INSTRUMENT_REQUIRED",
    MAX_DELEGATED_ADMINISTRATORS_FOR_SERVICE_LIMIT_EXCEEDED: "MAX_DELEGATED_ADMINISTRATORS_FOR_SERVICE_LIMIT_EXCEEDED",
    MAX_POLICY_TYPE_ATTACHMENT_LIMIT_EXCEEDED: "MAX_POLICY_TYPE_ATTACHMENT_LIMIT_EXCEEDED",
    MAX_TAG_LIMIT_EXCEEDED: "MAX_TAG_LIMIT_EXCEEDED",
    MEMBER_ACCOUNT_PAYMENT_INSTRUMENT_REQUIRED: "MEMBER_ACCOUNT_PAYMENT_INSTRUMENT_REQUIRED",
    MIN_POLICY_TYPE_ATTACHMENT_LIMIT_EXCEEDED: "MIN_POLICY_TYPE_ATTACHMENT_LIMIT_EXCEEDED",
    ORGANIZATION_NOT_IN_ALL_FEATURES_MODE: "ORGANIZATION_NOT_IN_ALL_FEATURES_MODE",
    OU_DEPTH_LIMIT_EXCEEDED: "OU_DEPTH_LIMIT_EXCEEDED",
    OU_NUMBER_LIMIT_EXCEEDED: "OU_NUMBER_LIMIT_EXCEEDED",
    POLICY_CONTENT_LIMIT_EXCEEDED: "POLICY_CONTENT_LIMIT_EXCEEDED",
    POLICY_NUMBER_LIMIT_EXCEEDED: "POLICY_NUMBER_LIMIT_EXCEEDED",
    POLICY_TYPE_ENABLED_FOR_THIS_SERVICE: "POLICY_TYPE_ENABLED_FOR_THIS_SERVICE",
    SERVICE_ACCESS_NOT_ENABLED: "SERVICE_ACCESS_NOT_ENABLED",
    TAG_POLICY_VIOLATION: "TAG_POLICY_VIOLATION",
    WAIT_PERIOD_ACTIVE: "WAIT_PERIOD_ACTIVE",
};
class ConstraintViolationException extends OrganizationsServiceException {
    name = "ConstraintViolationException";
    $fault = "client";
    Message;
    Reason;
    constructor(opts) {
        super({
            name: "ConstraintViolationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConstraintViolationException.prototype);
        this.Message = opts.Message;
        this.Reason = opts.Reason;
    }
}
class DuplicatePolicyAttachmentException extends OrganizationsServiceException {
    name = "DuplicatePolicyAttachmentException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "DuplicatePolicyAttachmentException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DuplicatePolicyAttachmentException.prototype);
        this.Message = opts.Message;
    }
}
class PolicyChangesInProgressException extends OrganizationsServiceException {
    name = "PolicyChangesInProgressException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "PolicyChangesInProgressException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PolicyChangesInProgressException.prototype);
        this.Message = opts.Message;
    }
}
class PolicyNotFoundException extends OrganizationsServiceException {
    name = "PolicyNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "PolicyNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PolicyNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class PolicyTypeNotEnabledException extends OrganizationsServiceException {
    name = "PolicyTypeNotEnabledException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "PolicyTypeNotEnabledException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PolicyTypeNotEnabledException.prototype);
        this.Message = opts.Message;
    }
}
class TargetNotFoundException extends OrganizationsServiceException {
    name = "TargetNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "TargetNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TargetNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class UnsupportedAPIEndpointException extends OrganizationsServiceException {
    name = "UnsupportedAPIEndpointException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "UnsupportedAPIEndpointException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, UnsupportedAPIEndpointException.prototype);
        this.Message = opts.Message;
    }
}
class ConflictException extends OrganizationsServiceException {
    name = "ConflictException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ConflictException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConflictException.prototype);
        this.Message = opts.Message;
    }
}
const IAMUserAccessToBilling = {
    ALLOW: "ALLOW",
    DENY: "DENY",
};
const CreateAccountFailureReason = {
    ACCOUNT_LIMIT_EXCEEDED: "ACCOUNT_LIMIT_EXCEEDED",
    CONCURRENT_ACCOUNT_MODIFICATION: "CONCURRENT_ACCOUNT_MODIFICATION",
    EMAIL_ALREADY_EXISTS: "EMAIL_ALREADY_EXISTS",
    FAILED_BUSINESS_VALIDATION: "FAILED_BUSINESS_VALIDATION",
    GOVCLOUD_ACCOUNT_ALREADY_EXISTS: "GOVCLOUD_ACCOUNT_ALREADY_EXISTS",
    INTERNAL_FAILURE: "INTERNAL_FAILURE",
    INVALID_ADDRESS: "INVALID_ADDRESS",
    INVALID_EMAIL: "INVALID_EMAIL",
    INVALID_IDENTITY_FOR_BUSINESS_VALIDATION: "INVALID_IDENTITY_FOR_BUSINESS_VALIDATION",
    INVALID_PAYMENT_INSTRUMENT: "INVALID_PAYMENT_INSTRUMENT",
    MISSING_BUSINESS_VALIDATION: "MISSING_BUSINESS_VALIDATION",
    MISSING_PAYMENT_INSTRUMENT: "MISSING_PAYMENT_INSTRUMENT",
    PENDING_BUSINESS_VALIDATIONv: "PENDING_BUSINESS_VALIDATION",
    UNKNOWN_BUSINESS_VALIDATION: "UNKNOWN_BUSINESS_VALIDATION",
    UPDATE_EXISTING_RESOURCE_POLICY_WITH_TAGS_NOT_SUPPORTED: "UPDATE_EXISTING_RESOURCE_POLICY_WITH_TAGS_NOT_SUPPORTED",
};
const CreateAccountState = {
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
    SUCCEEDED: "SUCCEEDED",
};
class FinalizingOrganizationException extends OrganizationsServiceException {
    name = "FinalizingOrganizationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "FinalizingOrganizationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, FinalizingOrganizationException.prototype);
        this.Message = opts.Message;
    }
}
const OrganizationFeatureSet = {
    ALL: "ALL",
    CONSOLIDATED_BILLING: "CONSOLIDATED_BILLING",
};
const PolicyTypeStatus = {
    ENABLED: "ENABLED",
    PENDING_DISABLE: "PENDING_DISABLE",
    PENDING_ENABLE: "PENDING_ENABLE",
};
const PolicyType = {
    AISERVICES_OPT_OUT_POLICY: "AISERVICES_OPT_OUT_POLICY",
    BACKUP_POLICY: "BACKUP_POLICY",
    CHATBOT_POLICY: "CHATBOT_POLICY",
    DECLARATIVE_POLICY_EC2: "DECLARATIVE_POLICY_EC2",
    RESOURCE_CONTROL_POLICY: "RESOURCE_CONTROL_POLICY",
    SECURITYHUB_POLICY: "SECURITYHUB_POLICY",
    SERVICE_CONTROL_POLICY: "SERVICE_CONTROL_POLICY",
    TAG_POLICY: "TAG_POLICY",
};
class DuplicateOrganizationalUnitException extends OrganizationsServiceException {
    name = "DuplicateOrganizationalUnitException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "DuplicateOrganizationalUnitException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DuplicateOrganizationalUnitException.prototype);
        this.Message = opts.Message;
    }
}
class ParentNotFoundException extends OrganizationsServiceException {
    name = "ParentNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ParentNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ParentNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class DuplicatePolicyException extends OrganizationsServiceException {
    name = "DuplicatePolicyException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "DuplicatePolicyException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DuplicatePolicyException.prototype);
        this.Message = opts.Message;
    }
}
class MalformedPolicyDocumentException extends OrganizationsServiceException {
    name = "MalformedPolicyDocumentException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "MalformedPolicyDocumentException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, MalformedPolicyDocumentException.prototype);
        this.Message = opts.Message;
    }
}
class PolicyTypeNotAvailableForOrganizationException extends OrganizationsServiceException {
    name = "PolicyTypeNotAvailableForOrganizationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "PolicyTypeNotAvailableForOrganizationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PolicyTypeNotAvailableForOrganizationException.prototype);
        this.Message = opts.Message;
    }
}
class OrganizationNotEmptyException extends OrganizationsServiceException {
    name = "OrganizationNotEmptyException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "OrganizationNotEmptyException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, OrganizationNotEmptyException.prototype);
        this.Message = opts.Message;
    }
}
class OrganizationalUnitNotEmptyException extends OrganizationsServiceException {
    name = "OrganizationalUnitNotEmptyException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "OrganizationalUnitNotEmptyException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, OrganizationalUnitNotEmptyException.prototype);
        this.Message = opts.Message;
    }
}
class OrganizationalUnitNotFoundException extends OrganizationsServiceException {
    name = "OrganizationalUnitNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "OrganizationalUnitNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, OrganizationalUnitNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class PolicyInUseException extends OrganizationsServiceException {
    name = "PolicyInUseException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "PolicyInUseException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PolicyInUseException.prototype);
        this.Message = opts.Message;
    }
}
class ResourcePolicyNotFoundException extends OrganizationsServiceException {
    name = "ResourcePolicyNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ResourcePolicyNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourcePolicyNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class CreateAccountStatusNotFoundException extends OrganizationsServiceException {
    name = "CreateAccountStatusNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "CreateAccountStatusNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, CreateAccountStatusNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
const EffectivePolicyType = {
    AISERVICES_OPT_OUT_POLICY: "AISERVICES_OPT_OUT_POLICY",
    BACKUP_POLICY: "BACKUP_POLICY",
    CHATBOT_POLICY: "CHATBOT_POLICY",
    DECLARATIVE_POLICY_EC2: "DECLARATIVE_POLICY_EC2",
    SECURITYHUB_POLICY: "SECURITYHUB_POLICY",
    TAG_POLICY: "TAG_POLICY",
};
class EffectivePolicyNotFoundException extends OrganizationsServiceException {
    name = "EffectivePolicyNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "EffectivePolicyNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, EffectivePolicyNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class PolicyNotAttachedException extends OrganizationsServiceException {
    name = "PolicyNotAttachedException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "PolicyNotAttachedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PolicyNotAttachedException.prototype);
        this.Message = opts.Message;
    }
}
class RootNotFoundException extends OrganizationsServiceException {
    name = "RootNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "RootNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, RootNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class PolicyTypeAlreadyEnabledException extends OrganizationsServiceException {
    name = "PolicyTypeAlreadyEnabledException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "PolicyTypeAlreadyEnabledException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PolicyTypeAlreadyEnabledException.prototype);
        this.Message = opts.Message;
    }
}
class DuplicateHandshakeException extends OrganizationsServiceException {
    name = "DuplicateHandshakeException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "DuplicateHandshakeException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DuplicateHandshakeException.prototype);
        this.Message = opts.Message;
    }
}
class MasterCannotLeaveOrganizationException extends OrganizationsServiceException {
    name = "MasterCannotLeaveOrganizationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "MasterCannotLeaveOrganizationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, MasterCannotLeaveOrganizationException.prototype);
        this.Message = opts.Message;
    }
}
const ChildType = {
    ACCOUNT: "ACCOUNT",
    ORGANIZATIONAL_UNIT: "ORGANIZATIONAL_UNIT",
};
class ChildNotFoundException extends OrganizationsServiceException {
    name = "ChildNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ChildNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ChildNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
const ParentType = {
    ORGANIZATIONAL_UNIT: "ORGANIZATIONAL_UNIT",
    ROOT: "ROOT",
};
const TargetType = {
    ACCOUNT: "ACCOUNT",
    ORGANIZATIONAL_UNIT: "ORGANIZATIONAL_UNIT",
    ROOT: "ROOT",
};
class DestinationParentNotFoundException extends OrganizationsServiceException {
    name = "DestinationParentNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "DestinationParentNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DestinationParentNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
class DuplicateAccountException extends OrganizationsServiceException {
    name = "DuplicateAccountException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "DuplicateAccountException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, DuplicateAccountException.prototype);
        this.Message = opts.Message;
    }
}
class SourceParentNotFoundException extends OrganizationsServiceException {
    name = "SourceParentNotFoundException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "SourceParentNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, SourceParentNotFoundException.prototype);
        this.Message = opts.Message;
    }
}
const HandshakePartyFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Id && { Id: smithyClient.SENSITIVE_STRING }),
});
const AccountFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Email && { Email: smithyClient.SENSITIVE_STRING }),
    ...(obj.Name && { Name: smithyClient.SENSITIVE_STRING }),
});
const CreateAccountRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Email && { Email: smithyClient.SENSITIVE_STRING }),
    ...(obj.AccountName && { AccountName: smithyClient.SENSITIVE_STRING }),
});
const CreateAccountStatusFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.AccountName && { AccountName: smithyClient.SENSITIVE_STRING }),
});
const CreateAccountResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.CreateAccountStatus && {
        CreateAccountStatus: CreateAccountStatusFilterSensitiveLog(obj.CreateAccountStatus),
    }),
});
const CreateGovCloudAccountRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Email && { Email: smithyClient.SENSITIVE_STRING }),
    ...(obj.AccountName && { AccountName: smithyClient.SENSITIVE_STRING }),
});
const CreateGovCloudAccountResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.CreateAccountStatus && {
        CreateAccountStatus: CreateAccountStatusFilterSensitiveLog(obj.CreateAccountStatus),
    }),
});
const OrganizationFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.MasterAccountEmail && { MasterAccountEmail: smithyClient.SENSITIVE_STRING }),
});
const CreateOrganizationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Organization && { Organization: OrganizationFilterSensitiveLog(obj.Organization) }),
});
const DescribeAccountResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Account && { Account: AccountFilterSensitiveLog(obj.Account) }),
});
const DescribeCreateAccountStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.CreateAccountStatus && {
        CreateAccountStatus: CreateAccountStatusFilterSensitiveLog(obj.CreateAccountStatus),
    }),
});
const DescribeOrganizationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Organization && { Organization: OrganizationFilterSensitiveLog(obj.Organization) }),
});
const InviteAccountToOrganizationRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Target && { Target: HandshakePartyFilterSensitiveLog(obj.Target) }),
    ...(obj.Notes && { Notes: smithyClient.SENSITIVE_STRING }),
});
const ListAccountsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Accounts && { Accounts: obj.Accounts.map((item) => AccountFilterSensitiveLog(item)) }),
});
const ListAccountsForParentResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Accounts && { Accounts: obj.Accounts.map((item) => AccountFilterSensitiveLog(item)) }),
});
const ListAccountsWithInvalidEffectivePolicyResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Accounts && { Accounts: obj.Accounts.map((item) => AccountFilterSensitiveLog(item)) }),
});
const ListCreateAccountStatusResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.CreateAccountStatuses && {
        CreateAccountStatuses: obj.CreateAccountStatuses.map((item) => CreateAccountStatusFilterSensitiveLog(item)),
    }),
});
const DelegatedAdministratorFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Email && { Email: smithyClient.SENSITIVE_STRING }),
    ...(obj.Name && { Name: smithyClient.SENSITIVE_STRING }),
});
const ListDelegatedAdministratorsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.DelegatedAdministrators && {
        DelegatedAdministrators: obj.DelegatedAdministrators.map((item) => DelegatedAdministratorFilterSensitiveLog(item)),
    }),
});
const HandshakeResourceFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Value && { Value: smithyClient.SENSITIVE_STRING }),
    ...(obj.Resources && { Resources: obj.Resources.map((item) => HandshakeResourceFilterSensitiveLog(item)) }),
});
const HandshakeFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Parties && { Parties: obj.Parties.map((item) => HandshakePartyFilterSensitiveLog(item)) }),
    ...(obj.Resources && { Resources: obj.Resources.map((item) => HandshakeResourceFilterSensitiveLog(item)) }),
});
const AcceptHandshakeResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Handshake && { Handshake: HandshakeFilterSensitiveLog(obj.Handshake) }),
});
const CancelHandshakeResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Handshake && { Handshake: HandshakeFilterSensitiveLog(obj.Handshake) }),
});
const DeclineHandshakeResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Handshake && { Handshake: HandshakeFilterSensitiveLog(obj.Handshake) }),
});
const DescribeHandshakeResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Handshake && { Handshake: HandshakeFilterSensitiveLog(obj.Handshake) }),
});
const EnableAllFeaturesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Handshake && { Handshake: HandshakeFilterSensitiveLog(obj.Handshake) }),
});
const InviteAccountToOrganizationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Handshake && { Handshake: HandshakeFilterSensitiveLog(obj.Handshake) }),
});
const ListHandshakesForAccountResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});
const ListHandshakesForOrganizationResponseFilterSensitiveLog = (obj) => ({
    ...obj,
});

const se_AcceptHandshakeCommand = async (input, context) => {
    const headers = sharedHeaders("AcceptHandshake");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_AttachPolicyCommand = async (input, context) => {
    const headers = sharedHeaders("AttachPolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_CancelHandshakeCommand = async (input, context) => {
    const headers = sharedHeaders("CancelHandshake");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_CloseAccountCommand = async (input, context) => {
    const headers = sharedHeaders("CloseAccount");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_CreateAccountCommand = async (input, context) => {
    const headers = sharedHeaders("CreateAccount");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_CreateGovCloudAccountCommand = async (input, context) => {
    const headers = sharedHeaders("CreateGovCloudAccount");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_CreateOrganizationCommand = async (input, context) => {
    const headers = sharedHeaders("CreateOrganization");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_CreateOrganizationalUnitCommand = async (input, context) => {
    const headers = sharedHeaders("CreateOrganizationalUnit");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_CreatePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("CreatePolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DeclineHandshakeCommand = async (input, context) => {
    const headers = sharedHeaders("DeclineHandshake");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DeleteOrganizationCommand = async (input, context) => {
    const headers = sharedHeaders("DeleteOrganization");
    const body = "{}";
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DeleteOrganizationalUnitCommand = async (input, context) => {
    const headers = sharedHeaders("DeleteOrganizationalUnit");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DeletePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("DeletePolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DeleteResourcePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("DeleteResourcePolicy");
    const body = "{}";
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DeregisterDelegatedAdministratorCommand = async (input, context) => {
    const headers = sharedHeaders("DeregisterDelegatedAdministrator");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribeAccountCommand = async (input, context) => {
    const headers = sharedHeaders("DescribeAccount");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribeCreateAccountStatusCommand = async (input, context) => {
    const headers = sharedHeaders("DescribeCreateAccountStatus");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribeEffectivePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("DescribeEffectivePolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribeHandshakeCommand = async (input, context) => {
    const headers = sharedHeaders("DescribeHandshake");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribeOrganizationCommand = async (input, context) => {
    const headers = sharedHeaders("DescribeOrganization");
    const body = "{}";
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribeOrganizationalUnitCommand = async (input, context) => {
    const headers = sharedHeaders("DescribeOrganizationalUnit");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("DescribePolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribeResourcePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("DescribeResourcePolicy");
    const body = "{}";
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DetachPolicyCommand = async (input, context) => {
    const headers = sharedHeaders("DetachPolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DisableAWSServiceAccessCommand = async (input, context) => {
    const headers = sharedHeaders("DisableAWSServiceAccess");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DisablePolicyTypeCommand = async (input, context) => {
    const headers = sharedHeaders("DisablePolicyType");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_EnableAllFeaturesCommand = async (input, context) => {
    const headers = sharedHeaders("EnableAllFeatures");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_EnableAWSServiceAccessCommand = async (input, context) => {
    const headers = sharedHeaders("EnableAWSServiceAccess");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_EnablePolicyTypeCommand = async (input, context) => {
    const headers = sharedHeaders("EnablePolicyType");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_InviteAccountToOrganizationCommand = async (input, context) => {
    const headers = sharedHeaders("InviteAccountToOrganization");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_LeaveOrganizationCommand = async (input, context) => {
    const headers = sharedHeaders("LeaveOrganization");
    const body = "{}";
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListAccountsCommand = async (input, context) => {
    const headers = sharedHeaders("ListAccounts");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListAccountsForParentCommand = async (input, context) => {
    const headers = sharedHeaders("ListAccountsForParent");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListAccountsWithInvalidEffectivePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("ListAccountsWithInvalidEffectivePolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListAWSServiceAccessForOrganizationCommand = async (input, context) => {
    const headers = sharedHeaders("ListAWSServiceAccessForOrganization");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListChildrenCommand = async (input, context) => {
    const headers = sharedHeaders("ListChildren");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListCreateAccountStatusCommand = async (input, context) => {
    const headers = sharedHeaders("ListCreateAccountStatus");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListDelegatedAdministratorsCommand = async (input, context) => {
    const headers = sharedHeaders("ListDelegatedAdministrators");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListDelegatedServicesForAccountCommand = async (input, context) => {
    const headers = sharedHeaders("ListDelegatedServicesForAccount");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListEffectivePolicyValidationErrorsCommand = async (input, context) => {
    const headers = sharedHeaders("ListEffectivePolicyValidationErrors");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListHandshakesForAccountCommand = async (input, context) => {
    const headers = sharedHeaders("ListHandshakesForAccount");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListHandshakesForOrganizationCommand = async (input, context) => {
    const headers = sharedHeaders("ListHandshakesForOrganization");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListOrganizationalUnitsForParentCommand = async (input, context) => {
    const headers = sharedHeaders("ListOrganizationalUnitsForParent");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListParentsCommand = async (input, context) => {
    const headers = sharedHeaders("ListParents");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListPoliciesCommand = async (input, context) => {
    const headers = sharedHeaders("ListPolicies");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListPoliciesForTargetCommand = async (input, context) => {
    const headers = sharedHeaders("ListPoliciesForTarget");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListRootsCommand = async (input, context) => {
    const headers = sharedHeaders("ListRoots");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListTagsForResourceCommand = async (input, context) => {
    const headers = sharedHeaders("ListTagsForResource");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListTargetsForPolicyCommand = async (input, context) => {
    const headers = sharedHeaders("ListTargetsForPolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_MoveAccountCommand = async (input, context) => {
    const headers = sharedHeaders("MoveAccount");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_PutResourcePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("PutResourcePolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_RegisterDelegatedAdministratorCommand = async (input, context) => {
    const headers = sharedHeaders("RegisterDelegatedAdministrator");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_RemoveAccountFromOrganizationCommand = async (input, context) => {
    const headers = sharedHeaders("RemoveAccountFromOrganization");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_TagResourceCommand = async (input, context) => {
    const headers = sharedHeaders("TagResource");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_UntagResourceCommand = async (input, context) => {
    const headers = sharedHeaders("UntagResource");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_UpdateOrganizationalUnitCommand = async (input, context) => {
    const headers = sharedHeaders("UpdateOrganizationalUnit");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_UpdatePolicyCommand = async (input, context) => {
    const headers = sharedHeaders("UpdatePolicy");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const de_AcceptHandshakeCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_AcceptHandshakeResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_AttachPolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_CancelHandshakeCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_CancelHandshakeResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_CloseAccountCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_CreateAccountCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_CreateAccountResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_CreateGovCloudAccountCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_CreateGovCloudAccountResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_CreateOrganizationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_CreateOrganizationalUnitCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_CreatePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DeclineHandshakeCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_DeclineHandshakeResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DeleteOrganizationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DeleteOrganizationalUnitCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DeletePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DeleteResourcePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DeregisterDelegatedAdministratorCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DescribeAccountCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_DescribeAccountResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DescribeCreateAccountStatusCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_DescribeCreateAccountStatusResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DescribeEffectivePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_DescribeEffectivePolicyResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DescribeHandshakeCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_DescribeHandshakeResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DescribeOrganizationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DescribeOrganizationalUnitCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DescribePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DescribeResourcePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_DetachPolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DisableAWSServiceAccessCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DisablePolicyTypeCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_EnableAllFeaturesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_EnableAllFeaturesResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_EnableAWSServiceAccessCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_EnablePolicyTypeCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_InviteAccountToOrganizationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_InviteAccountToOrganizationResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_LeaveOrganizationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_ListAccountsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListAccountsResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListAccountsForParentCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListAccountsForParentResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListAccountsWithInvalidEffectivePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListAccountsWithInvalidEffectivePolicyResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListAWSServiceAccessForOrganizationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListAWSServiceAccessForOrganizationResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListChildrenCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListCreateAccountStatusCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListCreateAccountStatusResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListDelegatedAdministratorsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListDelegatedAdministratorsResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListDelegatedServicesForAccountCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListDelegatedServicesForAccountResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListEffectivePolicyValidationErrorsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListEffectivePolicyValidationErrorsResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListHandshakesForAccountCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListHandshakesForAccountResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListHandshakesForOrganizationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListHandshakesForOrganizationResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListOrganizationalUnitsForParentCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListParentsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListPoliciesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListPoliciesForTargetCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListRootsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListTagsForResourceCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListTargetsForPolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_MoveAccountCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_PutResourcePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_RegisterDelegatedAdministratorCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_RemoveAccountFromOrganizationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_TagResourceCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_UntagResourceCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_UpdateOrganizationalUnitCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_UpdatePolicyCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_CommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await core$1.parseJsonErrorBody(output.body, context),
    };
    const errorCode = core$1.loadRestJsonErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "AWSOrganizationsNotInUseException":
        case "com.amazonaws.organizations#AWSOrganizationsNotInUseException":
            throw await de_AWSOrganizationsNotInUseExceptionRes(parsedOutput);
        case "AccessDeniedException":
        case "com.amazonaws.organizations#AccessDeniedException":
            throw await de_AccessDeniedExceptionRes(parsedOutput);
        case "AccessDeniedForDependencyException":
        case "com.amazonaws.organizations#AccessDeniedForDependencyException":
            throw await de_AccessDeniedForDependencyExceptionRes(parsedOutput);
        case "ConcurrentModificationException":
        case "com.amazonaws.organizations#ConcurrentModificationException":
            throw await de_ConcurrentModificationExceptionRes(parsedOutput);
        case "HandshakeAlreadyInStateException":
        case "com.amazonaws.organizations#HandshakeAlreadyInStateException":
            throw await de_HandshakeAlreadyInStateExceptionRes(parsedOutput);
        case "HandshakeConstraintViolationException":
        case "com.amazonaws.organizations#HandshakeConstraintViolationException":
            throw await de_HandshakeConstraintViolationExceptionRes(parsedOutput);
        case "HandshakeNotFoundException":
        case "com.amazonaws.organizations#HandshakeNotFoundException":
            throw await de_HandshakeNotFoundExceptionRes(parsedOutput);
        case "InvalidHandshakeTransitionException":
        case "com.amazonaws.organizations#InvalidHandshakeTransitionException":
            throw await de_InvalidHandshakeTransitionExceptionRes(parsedOutput);
        case "InvalidInputException":
        case "com.amazonaws.organizations#InvalidInputException":
            throw await de_InvalidInputExceptionRes(parsedOutput);
        case "ServiceException":
        case "com.amazonaws.organizations#ServiceException":
            throw await de_ServiceExceptionRes(parsedOutput);
        case "TooManyRequestsException":
        case "com.amazonaws.organizations#TooManyRequestsException":
            throw await de_TooManyRequestsExceptionRes(parsedOutput);
        case "ConstraintViolationException":
        case "com.amazonaws.organizations#ConstraintViolationException":
            throw await de_ConstraintViolationExceptionRes(parsedOutput);
        case "DuplicatePolicyAttachmentException":
        case "com.amazonaws.organizations#DuplicatePolicyAttachmentException":
            throw await de_DuplicatePolicyAttachmentExceptionRes(parsedOutput);
        case "PolicyChangesInProgressException":
        case "com.amazonaws.organizations#PolicyChangesInProgressException":
            throw await de_PolicyChangesInProgressExceptionRes(parsedOutput);
        case "PolicyNotFoundException":
        case "com.amazonaws.organizations#PolicyNotFoundException":
            throw await de_PolicyNotFoundExceptionRes(parsedOutput);
        case "PolicyTypeNotEnabledException":
        case "com.amazonaws.organizations#PolicyTypeNotEnabledException":
            throw await de_PolicyTypeNotEnabledExceptionRes(parsedOutput);
        case "TargetNotFoundException":
        case "com.amazonaws.organizations#TargetNotFoundException":
            throw await de_TargetNotFoundExceptionRes(parsedOutput);
        case "UnsupportedAPIEndpointException":
        case "com.amazonaws.organizations#UnsupportedAPIEndpointException":
            throw await de_UnsupportedAPIEndpointExceptionRes(parsedOutput);
        case "AccountAlreadyClosedException":
        case "com.amazonaws.organizations#AccountAlreadyClosedException":
            throw await de_AccountAlreadyClosedExceptionRes(parsedOutput);
        case "AccountNotFoundException":
        case "com.amazonaws.organizations#AccountNotFoundException":
            throw await de_AccountNotFoundExceptionRes(parsedOutput);
        case "ConflictException":
        case "com.amazonaws.organizations#ConflictException":
            throw await de_ConflictExceptionRes(parsedOutput);
        case "FinalizingOrganizationException":
        case "com.amazonaws.organizations#FinalizingOrganizationException":
            throw await de_FinalizingOrganizationExceptionRes(parsedOutput);
        case "AlreadyInOrganizationException":
        case "com.amazonaws.organizations#AlreadyInOrganizationException":
            throw await de_AlreadyInOrganizationExceptionRes(parsedOutput);
        case "DuplicateOrganizationalUnitException":
        case "com.amazonaws.organizations#DuplicateOrganizationalUnitException":
            throw await de_DuplicateOrganizationalUnitExceptionRes(parsedOutput);
        case "ParentNotFoundException":
        case "com.amazonaws.organizations#ParentNotFoundException":
            throw await de_ParentNotFoundExceptionRes(parsedOutput);
        case "DuplicatePolicyException":
        case "com.amazonaws.organizations#DuplicatePolicyException":
            throw await de_DuplicatePolicyExceptionRes(parsedOutput);
        case "MalformedPolicyDocumentException":
        case "com.amazonaws.organizations#MalformedPolicyDocumentException":
            throw await de_MalformedPolicyDocumentExceptionRes(parsedOutput);
        case "PolicyTypeNotAvailableForOrganizationException":
        case "com.amazonaws.organizations#PolicyTypeNotAvailableForOrganizationException":
            throw await de_PolicyTypeNotAvailableForOrganizationExceptionRes(parsedOutput);
        case "OrganizationNotEmptyException":
        case "com.amazonaws.organizations#OrganizationNotEmptyException":
            throw await de_OrganizationNotEmptyExceptionRes(parsedOutput);
        case "OrganizationalUnitNotEmptyException":
        case "com.amazonaws.organizations#OrganizationalUnitNotEmptyException":
            throw await de_OrganizationalUnitNotEmptyExceptionRes(parsedOutput);
        case "OrganizationalUnitNotFoundException":
        case "com.amazonaws.organizations#OrganizationalUnitNotFoundException":
            throw await de_OrganizationalUnitNotFoundExceptionRes(parsedOutput);
        case "PolicyInUseException":
        case "com.amazonaws.organizations#PolicyInUseException":
            throw await de_PolicyInUseExceptionRes(parsedOutput);
        case "ResourcePolicyNotFoundException":
        case "com.amazonaws.organizations#ResourcePolicyNotFoundException":
            throw await de_ResourcePolicyNotFoundExceptionRes(parsedOutput);
        case "AccountNotRegisteredException":
        case "com.amazonaws.organizations#AccountNotRegisteredException":
            throw await de_AccountNotRegisteredExceptionRes(parsedOutput);
        case "CreateAccountStatusNotFoundException":
        case "com.amazonaws.organizations#CreateAccountStatusNotFoundException":
            throw await de_CreateAccountStatusNotFoundExceptionRes(parsedOutput);
        case "EffectivePolicyNotFoundException":
        case "com.amazonaws.organizations#EffectivePolicyNotFoundException":
            throw await de_EffectivePolicyNotFoundExceptionRes(parsedOutput);
        case "PolicyNotAttachedException":
        case "com.amazonaws.organizations#PolicyNotAttachedException":
            throw await de_PolicyNotAttachedExceptionRes(parsedOutput);
        case "RootNotFoundException":
        case "com.amazonaws.organizations#RootNotFoundException":
            throw await de_RootNotFoundExceptionRes(parsedOutput);
        case "PolicyTypeAlreadyEnabledException":
        case "com.amazonaws.organizations#PolicyTypeAlreadyEnabledException":
            throw await de_PolicyTypeAlreadyEnabledExceptionRes(parsedOutput);
        case "AccountOwnerNotVerifiedException":
        case "com.amazonaws.organizations#AccountOwnerNotVerifiedException":
            throw await de_AccountOwnerNotVerifiedExceptionRes(parsedOutput);
        case "DuplicateHandshakeException":
        case "com.amazonaws.organizations#DuplicateHandshakeException":
            throw await de_DuplicateHandshakeExceptionRes(parsedOutput);
        case "MasterCannotLeaveOrganizationException":
        case "com.amazonaws.organizations#MasterCannotLeaveOrganizationException":
            throw await de_MasterCannotLeaveOrganizationExceptionRes(parsedOutput);
        case "ChildNotFoundException":
        case "com.amazonaws.organizations#ChildNotFoundException":
            throw await de_ChildNotFoundExceptionRes(parsedOutput);
        case "DestinationParentNotFoundException":
        case "com.amazonaws.organizations#DestinationParentNotFoundException":
            throw await de_DestinationParentNotFoundExceptionRes(parsedOutput);
        case "DuplicateAccountException":
        case "com.amazonaws.organizations#DuplicateAccountException":
            throw await de_DuplicateAccountExceptionRes(parsedOutput);
        case "SourceParentNotFoundException":
        case "com.amazonaws.organizations#SourceParentNotFoundException":
            throw await de_SourceParentNotFoundExceptionRes(parsedOutput);
        case "AccountAlreadyRegisteredException":
        case "com.amazonaws.organizations#AccountAlreadyRegisteredException":
            throw await de_AccountAlreadyRegisteredExceptionRes(parsedOutput);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody,
                errorCode,
            });
    }
};
const de_AccessDeniedExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AccessDeniedException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AccessDeniedForDependencyExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AccessDeniedForDependencyException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AccountAlreadyClosedExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AccountAlreadyClosedException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AccountAlreadyRegisteredExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AccountAlreadyRegisteredException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AccountNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AccountNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AccountNotRegisteredExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AccountNotRegisteredException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AccountOwnerNotVerifiedExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AccountOwnerNotVerifiedException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AlreadyInOrganizationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AlreadyInOrganizationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AWSOrganizationsNotInUseExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AWSOrganizationsNotInUseException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ChildNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ChildNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ConcurrentModificationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ConcurrentModificationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ConflictExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ConflictException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ConstraintViolationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ConstraintViolationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_CreateAccountStatusNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new CreateAccountStatusNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_DestinationParentNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new DestinationParentNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_DuplicateAccountExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new DuplicateAccountException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_DuplicateHandshakeExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new DuplicateHandshakeException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_DuplicateOrganizationalUnitExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new DuplicateOrganizationalUnitException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_DuplicatePolicyAttachmentExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new DuplicatePolicyAttachmentException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_DuplicatePolicyExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new DuplicatePolicyException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_EffectivePolicyNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new EffectivePolicyNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_FinalizingOrganizationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new FinalizingOrganizationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_HandshakeAlreadyInStateExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new HandshakeAlreadyInStateException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_HandshakeConstraintViolationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new HandshakeConstraintViolationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_HandshakeNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new HandshakeNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_InvalidHandshakeTransitionExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new InvalidHandshakeTransitionException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_InvalidInputExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new InvalidInputException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_MalformedPolicyDocumentExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new MalformedPolicyDocumentException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_MasterCannotLeaveOrganizationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new MasterCannotLeaveOrganizationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_OrganizationalUnitNotEmptyExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new OrganizationalUnitNotEmptyException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_OrganizationalUnitNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new OrganizationalUnitNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_OrganizationNotEmptyExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new OrganizationNotEmptyException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ParentNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ParentNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_PolicyChangesInProgressExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new PolicyChangesInProgressException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_PolicyInUseExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new PolicyInUseException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_PolicyNotAttachedExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new PolicyNotAttachedException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_PolicyNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new PolicyNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_PolicyTypeAlreadyEnabledExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new PolicyTypeAlreadyEnabledException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_PolicyTypeNotAvailableForOrganizationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new PolicyTypeNotAvailableForOrganizationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_PolicyTypeNotEnabledExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new PolicyTypeNotEnabledException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ResourcePolicyNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ResourcePolicyNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_RootNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new RootNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ServiceExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ServiceException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_SourceParentNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new SourceParentNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_TargetNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new TargetNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_TooManyRequestsExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new TooManyRequestsException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_UnsupportedAPIEndpointExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new UnsupportedAPIEndpointException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_AcceptHandshakeResponse = (output, context) => {
    return smithyClient.take(output, {
        Handshake: (_) => de_Handshake(_),
    });
};
const de_Account = (output, context) => {
    return smithyClient.take(output, {
        Arn: smithyClient.expectString,
        Email: smithyClient.expectString,
        Id: smithyClient.expectString,
        JoinedMethod: smithyClient.expectString,
        JoinedTimestamp: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Name: smithyClient.expectString,
        State: smithyClient.expectString,
        Status: smithyClient.expectString,
    });
};
const de_Accounts = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_Account(entry);
    });
    return retVal;
};
const de_CancelHandshakeResponse = (output, context) => {
    return smithyClient.take(output, {
        Handshake: (_) => de_Handshake(_),
    });
};
const de_CreateAccountResponse = (output, context) => {
    return smithyClient.take(output, {
        CreateAccountStatus: (_) => de_CreateAccountStatus(_),
    });
};
const de_CreateAccountStatus = (output, context) => {
    return smithyClient.take(output, {
        AccountId: smithyClient.expectString,
        AccountName: smithyClient.expectString,
        CompletedTimestamp: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        FailureReason: smithyClient.expectString,
        GovCloudAccountId: smithyClient.expectString,
        Id: smithyClient.expectString,
        RequestedTimestamp: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        State: smithyClient.expectString,
    });
};
const de_CreateAccountStatuses = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_CreateAccountStatus(entry);
    });
    return retVal;
};
const de_CreateGovCloudAccountResponse = (output, context) => {
    return smithyClient.take(output, {
        CreateAccountStatus: (_) => de_CreateAccountStatus(_),
    });
};
const de_DeclineHandshakeResponse = (output, context) => {
    return smithyClient.take(output, {
        Handshake: (_) => de_Handshake(_),
    });
};
const de_DelegatedAdministrator = (output, context) => {
    return smithyClient.take(output, {
        Arn: smithyClient.expectString,
        DelegationEnabledDate: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Email: smithyClient.expectString,
        Id: smithyClient.expectString,
        JoinedMethod: smithyClient.expectString,
        JoinedTimestamp: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Name: smithyClient.expectString,
        State: smithyClient.expectString,
        Status: smithyClient.expectString,
    });
};
const de_DelegatedAdministrators = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_DelegatedAdministrator(entry);
    });
    return retVal;
};
const de_DelegatedService = (output, context) => {
    return smithyClient.take(output, {
        DelegationEnabledDate: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        ServicePrincipal: smithyClient.expectString,
    });
};
const de_DelegatedServices = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_DelegatedService(entry);
    });
    return retVal;
};
const de_DescribeAccountResponse = (output, context) => {
    return smithyClient.take(output, {
        Account: (_) => de_Account(_),
    });
};
const de_DescribeCreateAccountStatusResponse = (output, context) => {
    return smithyClient.take(output, {
        CreateAccountStatus: (_) => de_CreateAccountStatus(_),
    });
};
const de_DescribeEffectivePolicyResponse = (output, context) => {
    return smithyClient.take(output, {
        EffectivePolicy: (_) => de_EffectivePolicy(_),
    });
};
const de_DescribeHandshakeResponse = (output, context) => {
    return smithyClient.take(output, {
        Handshake: (_) => de_Handshake(_),
    });
};
const de_EffectivePolicy = (output, context) => {
    return smithyClient.take(output, {
        LastUpdatedTimestamp: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        PolicyContent: smithyClient.expectString,
        PolicyType: smithyClient.expectString,
        TargetId: smithyClient.expectString,
    });
};
const de_EnableAllFeaturesResponse = (output, context) => {
    return smithyClient.take(output, {
        Handshake: (_) => de_Handshake(_),
    });
};
const de_EnabledServicePrincipal = (output, context) => {
    return smithyClient.take(output, {
        DateEnabled: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        ServicePrincipal: smithyClient.expectString,
    });
};
const de_EnabledServicePrincipals = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_EnabledServicePrincipal(entry);
    });
    return retVal;
};
const de_Handshake = (output, context) => {
    return smithyClient.take(output, {
        Action: smithyClient.expectString,
        Arn: smithyClient.expectString,
        ExpirationTimestamp: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Id: smithyClient.expectString,
        Parties: smithyClient._json,
        RequestedTimestamp: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Resources: (_) => de_HandshakeResources(_),
        State: smithyClient.expectString,
    });
};
const de_HandshakeResource = (output, context) => {
    return smithyClient.take(output, {
        Resources: (_) => de_HandshakeResources(_),
        Type: smithyClient.expectString,
        Value: smithyClient.expectString,
    });
};
const de_HandshakeResources = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_HandshakeResource(entry);
    });
    return retVal;
};
const de_Handshakes = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_Handshake(entry);
    });
    return retVal;
};
const de_InviteAccountToOrganizationResponse = (output, context) => {
    return smithyClient.take(output, {
        Handshake: (_) => de_Handshake(_),
    });
};
const de_ListAccountsForParentResponse = (output, context) => {
    return smithyClient.take(output, {
        Accounts: (_) => de_Accounts(_),
        NextToken: smithyClient.expectString,
    });
};
const de_ListAccountsResponse = (output, context) => {
    return smithyClient.take(output, {
        Accounts: (_) => de_Accounts(_),
        NextToken: smithyClient.expectString,
    });
};
const de_ListAccountsWithInvalidEffectivePolicyResponse = (output, context) => {
    return smithyClient.take(output, {
        Accounts: (_) => de_Accounts(_),
        NextToken: smithyClient.expectString,
        PolicyType: smithyClient.expectString,
    });
};
const de_ListAWSServiceAccessForOrganizationResponse = (output, context) => {
    return smithyClient.take(output, {
        EnabledServicePrincipals: (_) => de_EnabledServicePrincipals(_),
        NextToken: smithyClient.expectString,
    });
};
const de_ListCreateAccountStatusResponse = (output, context) => {
    return smithyClient.take(output, {
        CreateAccountStatuses: (_) => de_CreateAccountStatuses(_),
        NextToken: smithyClient.expectString,
    });
};
const de_ListDelegatedAdministratorsResponse = (output, context) => {
    return smithyClient.take(output, {
        DelegatedAdministrators: (_) => de_DelegatedAdministrators(_),
        NextToken: smithyClient.expectString,
    });
};
const de_ListDelegatedServicesForAccountResponse = (output, context) => {
    return smithyClient.take(output, {
        DelegatedServices: (_) => de_DelegatedServices(_),
        NextToken: smithyClient.expectString,
    });
};
const de_ListEffectivePolicyValidationErrorsResponse = (output, context) => {
    return smithyClient.take(output, {
        AccountId: smithyClient.expectString,
        EffectivePolicyValidationErrors: smithyClient._json,
        EvaluationTimestamp: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        NextToken: smithyClient.expectString,
        Path: smithyClient.expectString,
        PolicyType: smithyClient.expectString,
    });
};
const de_ListHandshakesForAccountResponse = (output, context) => {
    return smithyClient.take(output, {
        Handshakes: (_) => de_Handshakes(_),
        NextToken: smithyClient.expectString,
    });
};
const de_ListHandshakesForOrganizationResponse = (output, context) => {
    return smithyClient.take(output, {
        Handshakes: (_) => de_Handshakes(_),
        NextToken: smithyClient.expectString,
    });
};
const deserializeMetadata = (output) => ({
    httpStatusCode: output.statusCode,
    requestId: output.headers["x-amzn-requestid"] ?? output.headers["x-amzn-request-id"] ?? output.headers["x-amz-request-id"],
    extendedRequestId: output.headers["x-amz-id-2"],
    cfId: output.headers["x-amz-cf-id"],
});
const throwDefaultError = smithyClient.withBaseException(OrganizationsServiceException);
const buildHttpRpcRequest = async (context, headers, path, resolvedHostname, body) => {
    const { hostname, protocol = "https", port, path: basePath } = await context.endpoint();
    const contents = {
        protocol,
        hostname,
        port,
        method: "POST",
        path: basePath.endsWith("/") ? basePath.slice(0, -1) + path : basePath + path,
        headers,
    };
    if (body !== undefined) {
        contents.body = body;
    }
    return new protocolHttp.HttpRequest(contents);
};
function sharedHeaders(operation) {
    return {
        "content-type": "application/x-amz-json-1.1",
        "x-amz-target": `AWSOrganizationsV20161128.${operation}`,
    };
}

class AcceptHandshakeCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "AcceptHandshake", {})
    .n("OrganizationsClient", "AcceptHandshakeCommand")
    .f(void 0, AcceptHandshakeResponseFilterSensitiveLog)
    .ser(se_AcceptHandshakeCommand)
    .de(de_AcceptHandshakeCommand)
    .build() {
}

class AttachPolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "AttachPolicy", {})
    .n("OrganizationsClient", "AttachPolicyCommand")
    .f(void 0, void 0)
    .ser(se_AttachPolicyCommand)
    .de(de_AttachPolicyCommand)
    .build() {
}

class CancelHandshakeCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "CancelHandshake", {})
    .n("OrganizationsClient", "CancelHandshakeCommand")
    .f(void 0, CancelHandshakeResponseFilterSensitiveLog)
    .ser(se_CancelHandshakeCommand)
    .de(de_CancelHandshakeCommand)
    .build() {
}

class CloseAccountCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "CloseAccount", {})
    .n("OrganizationsClient", "CloseAccountCommand")
    .f(void 0, void 0)
    .ser(se_CloseAccountCommand)
    .de(de_CloseAccountCommand)
    .build() {
}

class CreateAccountCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "CreateAccount", {})
    .n("OrganizationsClient", "CreateAccountCommand")
    .f(CreateAccountRequestFilterSensitiveLog, CreateAccountResponseFilterSensitiveLog)
    .ser(se_CreateAccountCommand)
    .de(de_CreateAccountCommand)
    .build() {
}

class CreateGovCloudAccountCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "CreateGovCloudAccount", {})
    .n("OrganizationsClient", "CreateGovCloudAccountCommand")
    .f(CreateGovCloudAccountRequestFilterSensitiveLog, CreateGovCloudAccountResponseFilterSensitiveLog)
    .ser(se_CreateGovCloudAccountCommand)
    .de(de_CreateGovCloudAccountCommand)
    .build() {
}

class CreateOrganizationalUnitCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "CreateOrganizationalUnit", {})
    .n("OrganizationsClient", "CreateOrganizationalUnitCommand")
    .f(void 0, void 0)
    .ser(se_CreateOrganizationalUnitCommand)
    .de(de_CreateOrganizationalUnitCommand)
    .build() {
}

class CreateOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "CreateOrganization", {})
    .n("OrganizationsClient", "CreateOrganizationCommand")
    .f(void 0, CreateOrganizationResponseFilterSensitiveLog)
    .ser(se_CreateOrganizationCommand)
    .de(de_CreateOrganizationCommand)
    .build() {
}

class CreatePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "CreatePolicy", {})
    .n("OrganizationsClient", "CreatePolicyCommand")
    .f(void 0, void 0)
    .ser(se_CreatePolicyCommand)
    .de(de_CreatePolicyCommand)
    .build() {
}

class DeclineHandshakeCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DeclineHandshake", {})
    .n("OrganizationsClient", "DeclineHandshakeCommand")
    .f(void 0, DeclineHandshakeResponseFilterSensitiveLog)
    .ser(se_DeclineHandshakeCommand)
    .de(de_DeclineHandshakeCommand)
    .build() {
}

class DeleteOrganizationalUnitCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DeleteOrganizationalUnit", {})
    .n("OrganizationsClient", "DeleteOrganizationalUnitCommand")
    .f(void 0, void 0)
    .ser(se_DeleteOrganizationalUnitCommand)
    .de(de_DeleteOrganizationalUnitCommand)
    .build() {
}

class DeleteOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DeleteOrganization", {})
    .n("OrganizationsClient", "DeleteOrganizationCommand")
    .f(void 0, void 0)
    .ser(se_DeleteOrganizationCommand)
    .de(de_DeleteOrganizationCommand)
    .build() {
}

class DeletePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DeletePolicy", {})
    .n("OrganizationsClient", "DeletePolicyCommand")
    .f(void 0, void 0)
    .ser(se_DeletePolicyCommand)
    .de(de_DeletePolicyCommand)
    .build() {
}

class DeleteResourcePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DeleteResourcePolicy", {})
    .n("OrganizationsClient", "DeleteResourcePolicyCommand")
    .f(void 0, void 0)
    .ser(se_DeleteResourcePolicyCommand)
    .de(de_DeleteResourcePolicyCommand)
    .build() {
}

class DeregisterDelegatedAdministratorCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DeregisterDelegatedAdministrator", {})
    .n("OrganizationsClient", "DeregisterDelegatedAdministratorCommand")
    .f(void 0, void 0)
    .ser(se_DeregisterDelegatedAdministratorCommand)
    .de(de_DeregisterDelegatedAdministratorCommand)
    .build() {
}

class DescribeAccountCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DescribeAccount", {})
    .n("OrganizationsClient", "DescribeAccountCommand")
    .f(void 0, DescribeAccountResponseFilterSensitiveLog)
    .ser(se_DescribeAccountCommand)
    .de(de_DescribeAccountCommand)
    .build() {
}

class DescribeCreateAccountStatusCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DescribeCreateAccountStatus", {})
    .n("OrganizationsClient", "DescribeCreateAccountStatusCommand")
    .f(void 0, DescribeCreateAccountStatusResponseFilterSensitiveLog)
    .ser(se_DescribeCreateAccountStatusCommand)
    .de(de_DescribeCreateAccountStatusCommand)
    .build() {
}

class DescribeEffectivePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DescribeEffectivePolicy", {})
    .n("OrganizationsClient", "DescribeEffectivePolicyCommand")
    .f(void 0, void 0)
    .ser(se_DescribeEffectivePolicyCommand)
    .de(de_DescribeEffectivePolicyCommand)
    .build() {
}

class DescribeHandshakeCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DescribeHandshake", {})
    .n("OrganizationsClient", "DescribeHandshakeCommand")
    .f(void 0, DescribeHandshakeResponseFilterSensitiveLog)
    .ser(se_DescribeHandshakeCommand)
    .de(de_DescribeHandshakeCommand)
    .build() {
}

class DescribeOrganizationalUnitCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DescribeOrganizationalUnit", {})
    .n("OrganizationsClient", "DescribeOrganizationalUnitCommand")
    .f(void 0, void 0)
    .ser(se_DescribeOrganizationalUnitCommand)
    .de(de_DescribeOrganizationalUnitCommand)
    .build() {
}

class DescribeOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DescribeOrganization", {})
    .n("OrganizationsClient", "DescribeOrganizationCommand")
    .f(void 0, DescribeOrganizationResponseFilterSensitiveLog)
    .ser(se_DescribeOrganizationCommand)
    .de(de_DescribeOrganizationCommand)
    .build() {
}

class DescribePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DescribePolicy", {})
    .n("OrganizationsClient", "DescribePolicyCommand")
    .f(void 0, void 0)
    .ser(se_DescribePolicyCommand)
    .de(de_DescribePolicyCommand)
    .build() {
}

class DescribeResourcePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DescribeResourcePolicy", {})
    .n("OrganizationsClient", "DescribeResourcePolicyCommand")
    .f(void 0, void 0)
    .ser(se_DescribeResourcePolicyCommand)
    .de(de_DescribeResourcePolicyCommand)
    .build() {
}

class DetachPolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DetachPolicy", {})
    .n("OrganizationsClient", "DetachPolicyCommand")
    .f(void 0, void 0)
    .ser(se_DetachPolicyCommand)
    .de(de_DetachPolicyCommand)
    .build() {
}

class DisableAWSServiceAccessCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DisableAWSServiceAccess", {})
    .n("OrganizationsClient", "DisableAWSServiceAccessCommand")
    .f(void 0, void 0)
    .ser(se_DisableAWSServiceAccessCommand)
    .de(de_DisableAWSServiceAccessCommand)
    .build() {
}

class DisablePolicyTypeCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "DisablePolicyType", {})
    .n("OrganizationsClient", "DisablePolicyTypeCommand")
    .f(void 0, void 0)
    .ser(se_DisablePolicyTypeCommand)
    .de(de_DisablePolicyTypeCommand)
    .build() {
}

class EnableAllFeaturesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "EnableAllFeatures", {})
    .n("OrganizationsClient", "EnableAllFeaturesCommand")
    .f(void 0, EnableAllFeaturesResponseFilterSensitiveLog)
    .ser(se_EnableAllFeaturesCommand)
    .de(de_EnableAllFeaturesCommand)
    .build() {
}

class EnableAWSServiceAccessCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "EnableAWSServiceAccess", {})
    .n("OrganizationsClient", "EnableAWSServiceAccessCommand")
    .f(void 0, void 0)
    .ser(se_EnableAWSServiceAccessCommand)
    .de(de_EnableAWSServiceAccessCommand)
    .build() {
}

class EnablePolicyTypeCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "EnablePolicyType", {})
    .n("OrganizationsClient", "EnablePolicyTypeCommand")
    .f(void 0, void 0)
    .ser(se_EnablePolicyTypeCommand)
    .de(de_EnablePolicyTypeCommand)
    .build() {
}

class InviteAccountToOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "InviteAccountToOrganization", {})
    .n("OrganizationsClient", "InviteAccountToOrganizationCommand")
    .f(InviteAccountToOrganizationRequestFilterSensitiveLog, InviteAccountToOrganizationResponseFilterSensitiveLog)
    .ser(se_InviteAccountToOrganizationCommand)
    .de(de_InviteAccountToOrganizationCommand)
    .build() {
}

class LeaveOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "LeaveOrganization", {})
    .n("OrganizationsClient", "LeaveOrganizationCommand")
    .f(void 0, void 0)
    .ser(se_LeaveOrganizationCommand)
    .de(de_LeaveOrganizationCommand)
    .build() {
}

class ListAccountsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListAccounts", {})
    .n("OrganizationsClient", "ListAccountsCommand")
    .f(void 0, ListAccountsResponseFilterSensitiveLog)
    .ser(se_ListAccountsCommand)
    .de(de_ListAccountsCommand)
    .build() {
}

class ListAccountsForParentCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListAccountsForParent", {})
    .n("OrganizationsClient", "ListAccountsForParentCommand")
    .f(void 0, ListAccountsForParentResponseFilterSensitiveLog)
    .ser(se_ListAccountsForParentCommand)
    .de(de_ListAccountsForParentCommand)
    .build() {
}

class ListAccountsWithInvalidEffectivePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListAccountsWithInvalidEffectivePolicy", {})
    .n("OrganizationsClient", "ListAccountsWithInvalidEffectivePolicyCommand")
    .f(void 0, ListAccountsWithInvalidEffectivePolicyResponseFilterSensitiveLog)
    .ser(se_ListAccountsWithInvalidEffectivePolicyCommand)
    .de(de_ListAccountsWithInvalidEffectivePolicyCommand)
    .build() {
}

class ListAWSServiceAccessForOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListAWSServiceAccessForOrganization", {})
    .n("OrganizationsClient", "ListAWSServiceAccessForOrganizationCommand")
    .f(void 0, void 0)
    .ser(se_ListAWSServiceAccessForOrganizationCommand)
    .de(de_ListAWSServiceAccessForOrganizationCommand)
    .build() {
}

class ListChildrenCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListChildren", {})
    .n("OrganizationsClient", "ListChildrenCommand")
    .f(void 0, void 0)
    .ser(se_ListChildrenCommand)
    .de(de_ListChildrenCommand)
    .build() {
}

class ListCreateAccountStatusCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListCreateAccountStatus", {})
    .n("OrganizationsClient", "ListCreateAccountStatusCommand")
    .f(void 0, ListCreateAccountStatusResponseFilterSensitiveLog)
    .ser(se_ListCreateAccountStatusCommand)
    .de(de_ListCreateAccountStatusCommand)
    .build() {
}

class ListDelegatedAdministratorsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListDelegatedAdministrators", {})
    .n("OrganizationsClient", "ListDelegatedAdministratorsCommand")
    .f(void 0, ListDelegatedAdministratorsResponseFilterSensitiveLog)
    .ser(se_ListDelegatedAdministratorsCommand)
    .de(de_ListDelegatedAdministratorsCommand)
    .build() {
}

class ListDelegatedServicesForAccountCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListDelegatedServicesForAccount", {})
    .n("OrganizationsClient", "ListDelegatedServicesForAccountCommand")
    .f(void 0, void 0)
    .ser(se_ListDelegatedServicesForAccountCommand)
    .de(de_ListDelegatedServicesForAccountCommand)
    .build() {
}

class ListEffectivePolicyValidationErrorsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListEffectivePolicyValidationErrors", {})
    .n("OrganizationsClient", "ListEffectivePolicyValidationErrorsCommand")
    .f(void 0, void 0)
    .ser(se_ListEffectivePolicyValidationErrorsCommand)
    .de(de_ListEffectivePolicyValidationErrorsCommand)
    .build() {
}

class ListHandshakesForAccountCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListHandshakesForAccount", {})
    .n("OrganizationsClient", "ListHandshakesForAccountCommand")
    .f(void 0, ListHandshakesForAccountResponseFilterSensitiveLog)
    .ser(se_ListHandshakesForAccountCommand)
    .de(de_ListHandshakesForAccountCommand)
    .build() {
}

class ListHandshakesForOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListHandshakesForOrganization", {})
    .n("OrganizationsClient", "ListHandshakesForOrganizationCommand")
    .f(void 0, ListHandshakesForOrganizationResponseFilterSensitiveLog)
    .ser(se_ListHandshakesForOrganizationCommand)
    .de(de_ListHandshakesForOrganizationCommand)
    .build() {
}

class ListOrganizationalUnitsForParentCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListOrganizationalUnitsForParent", {})
    .n("OrganizationsClient", "ListOrganizationalUnitsForParentCommand")
    .f(void 0, void 0)
    .ser(se_ListOrganizationalUnitsForParentCommand)
    .de(de_ListOrganizationalUnitsForParentCommand)
    .build() {
}

class ListParentsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListParents", {})
    .n("OrganizationsClient", "ListParentsCommand")
    .f(void 0, void 0)
    .ser(se_ListParentsCommand)
    .de(de_ListParentsCommand)
    .build() {
}

class ListPoliciesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListPolicies", {})
    .n("OrganizationsClient", "ListPoliciesCommand")
    .f(void 0, void 0)
    .ser(se_ListPoliciesCommand)
    .de(de_ListPoliciesCommand)
    .build() {
}

class ListPoliciesForTargetCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListPoliciesForTarget", {})
    .n("OrganizationsClient", "ListPoliciesForTargetCommand")
    .f(void 0, void 0)
    .ser(se_ListPoliciesForTargetCommand)
    .de(de_ListPoliciesForTargetCommand)
    .build() {
}

class ListRootsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListRoots", {})
    .n("OrganizationsClient", "ListRootsCommand")
    .f(void 0, void 0)
    .ser(se_ListRootsCommand)
    .de(de_ListRootsCommand)
    .build() {
}

class ListTagsForResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListTagsForResource", {})
    .n("OrganizationsClient", "ListTagsForResourceCommand")
    .f(void 0, void 0)
    .ser(se_ListTagsForResourceCommand)
    .de(de_ListTagsForResourceCommand)
    .build() {
}

class ListTargetsForPolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "ListTargetsForPolicy", {})
    .n("OrganizationsClient", "ListTargetsForPolicyCommand")
    .f(void 0, void 0)
    .ser(se_ListTargetsForPolicyCommand)
    .de(de_ListTargetsForPolicyCommand)
    .build() {
}

class MoveAccountCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "MoveAccount", {})
    .n("OrganizationsClient", "MoveAccountCommand")
    .f(void 0, void 0)
    .ser(se_MoveAccountCommand)
    .de(de_MoveAccountCommand)
    .build() {
}

class PutResourcePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "PutResourcePolicy", {})
    .n("OrganizationsClient", "PutResourcePolicyCommand")
    .f(void 0, void 0)
    .ser(se_PutResourcePolicyCommand)
    .de(de_PutResourcePolicyCommand)
    .build() {
}

class RegisterDelegatedAdministratorCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "RegisterDelegatedAdministrator", {})
    .n("OrganizationsClient", "RegisterDelegatedAdministratorCommand")
    .f(void 0, void 0)
    .ser(se_RegisterDelegatedAdministratorCommand)
    .de(de_RegisterDelegatedAdministratorCommand)
    .build() {
}

class RemoveAccountFromOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "RemoveAccountFromOrganization", {})
    .n("OrganizationsClient", "RemoveAccountFromOrganizationCommand")
    .f(void 0, void 0)
    .ser(se_RemoveAccountFromOrganizationCommand)
    .de(de_RemoveAccountFromOrganizationCommand)
    .build() {
}

class TagResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "TagResource", {})
    .n("OrganizationsClient", "TagResourceCommand")
    .f(void 0, void 0)
    .ser(se_TagResourceCommand)
    .de(de_TagResourceCommand)
    .build() {
}

class UntagResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "UntagResource", {})
    .n("OrganizationsClient", "UntagResourceCommand")
    .f(void 0, void 0)
    .ser(se_UntagResourceCommand)
    .de(de_UntagResourceCommand)
    .build() {
}

class UpdateOrganizationalUnitCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "UpdateOrganizationalUnit", {})
    .n("OrganizationsClient", "UpdateOrganizationalUnitCommand")
    .f(void 0, void 0)
    .ser(se_UpdateOrganizationalUnitCommand)
    .de(de_UpdateOrganizationalUnitCommand)
    .build() {
}

class UpdatePolicyCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSOrganizationsV20161128", "UpdatePolicy", {})
    .n("OrganizationsClient", "UpdatePolicyCommand")
    .f(void 0, void 0)
    .ser(se_UpdatePolicyCommand)
    .de(de_UpdatePolicyCommand)
    .build() {
}

const commands = {
    AcceptHandshakeCommand,
    AttachPolicyCommand,
    CancelHandshakeCommand,
    CloseAccountCommand,
    CreateAccountCommand,
    CreateGovCloudAccountCommand,
    CreateOrganizationCommand,
    CreateOrganizationalUnitCommand,
    CreatePolicyCommand,
    DeclineHandshakeCommand,
    DeleteOrganizationCommand,
    DeleteOrganizationalUnitCommand,
    DeletePolicyCommand,
    DeleteResourcePolicyCommand,
    DeregisterDelegatedAdministratorCommand,
    DescribeAccountCommand,
    DescribeCreateAccountStatusCommand,
    DescribeEffectivePolicyCommand,
    DescribeHandshakeCommand,
    DescribeOrganizationCommand,
    DescribeOrganizationalUnitCommand,
    DescribePolicyCommand,
    DescribeResourcePolicyCommand,
    DetachPolicyCommand,
    DisableAWSServiceAccessCommand,
    DisablePolicyTypeCommand,
    EnableAllFeaturesCommand,
    EnableAWSServiceAccessCommand,
    EnablePolicyTypeCommand,
    InviteAccountToOrganizationCommand,
    LeaveOrganizationCommand,
    ListAccountsCommand,
    ListAccountsForParentCommand,
    ListAccountsWithInvalidEffectivePolicyCommand,
    ListAWSServiceAccessForOrganizationCommand,
    ListChildrenCommand,
    ListCreateAccountStatusCommand,
    ListDelegatedAdministratorsCommand,
    ListDelegatedServicesForAccountCommand,
    ListEffectivePolicyValidationErrorsCommand,
    ListHandshakesForAccountCommand,
    ListHandshakesForOrganizationCommand,
    ListOrganizationalUnitsForParentCommand,
    ListParentsCommand,
    ListPoliciesCommand,
    ListPoliciesForTargetCommand,
    ListRootsCommand,
    ListTagsForResourceCommand,
    ListTargetsForPolicyCommand,
    MoveAccountCommand,
    PutResourcePolicyCommand,
    RegisterDelegatedAdministratorCommand,
    RemoveAccountFromOrganizationCommand,
    TagResourceCommand,
    UntagResourceCommand,
    UpdateOrganizationalUnitCommand,
    UpdatePolicyCommand,
};
class Organizations extends OrganizationsClient {
}
smithyClient.createAggregatedClient(commands, Organizations);

const paginateListAWSServiceAccessForOrganization = core.createPaginator(OrganizationsClient, ListAWSServiceAccessForOrganizationCommand, "NextToken", "NextToken", "MaxResults");

const paginateListAccountsForParent = core.createPaginator(OrganizationsClient, ListAccountsForParentCommand, "NextToken", "NextToken", "MaxResults");

const paginateListAccounts = core.createPaginator(OrganizationsClient, ListAccountsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListAccountsWithInvalidEffectivePolicy = core.createPaginator(OrganizationsClient, ListAccountsWithInvalidEffectivePolicyCommand, "NextToken", "NextToken", "MaxResults");

const paginateListChildren = core.createPaginator(OrganizationsClient, ListChildrenCommand, "NextToken", "NextToken", "MaxResults");

const paginateListCreateAccountStatus = core.createPaginator(OrganizationsClient, ListCreateAccountStatusCommand, "NextToken", "NextToken", "MaxResults");

const paginateListDelegatedAdministrators = core.createPaginator(OrganizationsClient, ListDelegatedAdministratorsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListDelegatedServicesForAccount = core.createPaginator(OrganizationsClient, ListDelegatedServicesForAccountCommand, "NextToken", "NextToken", "MaxResults");

const paginateListEffectivePolicyValidationErrors = core.createPaginator(OrganizationsClient, ListEffectivePolicyValidationErrorsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListHandshakesForAccount = core.createPaginator(OrganizationsClient, ListHandshakesForAccountCommand, "NextToken", "NextToken", "MaxResults");

const paginateListHandshakesForOrganization = core.createPaginator(OrganizationsClient, ListHandshakesForOrganizationCommand, "NextToken", "NextToken", "MaxResults");

const paginateListOrganizationalUnitsForParent = core.createPaginator(OrganizationsClient, ListOrganizationalUnitsForParentCommand, "NextToken", "NextToken", "MaxResults");

const paginateListParents = core.createPaginator(OrganizationsClient, ListParentsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListPoliciesForTarget = core.createPaginator(OrganizationsClient, ListPoliciesForTargetCommand, "NextToken", "NextToken", "MaxResults");

const paginateListPolicies = core.createPaginator(OrganizationsClient, ListPoliciesCommand, "NextToken", "NextToken", "MaxResults");

const paginateListRoots = core.createPaginator(OrganizationsClient, ListRootsCommand, "NextToken", "NextToken", "MaxResults");

const paginateListTagsForResource = core.createPaginator(OrganizationsClient, ListTagsForResourceCommand, "NextToken", "NextToken", "");

const paginateListTargetsForPolicy = core.createPaginator(OrganizationsClient, ListTargetsForPolicyCommand, "NextToken", "NextToken", "MaxResults");

Object.defineProperty(exports, "$Command", {
    enumerable: true,
    get: function () { return smithyClient.Command; }
});
Object.defineProperty(exports, "__Client", {
    enumerable: true,
    get: function () { return smithyClient.Client; }
});
exports.AWSOrganizationsNotInUseException = AWSOrganizationsNotInUseException;
exports.AcceptHandshakeCommand = AcceptHandshakeCommand;
exports.AcceptHandshakeResponseFilterSensitiveLog = AcceptHandshakeResponseFilterSensitiveLog;
exports.AccessDeniedException = AccessDeniedException;
exports.AccessDeniedForDependencyException = AccessDeniedForDependencyException;
exports.AccessDeniedForDependencyExceptionReason = AccessDeniedForDependencyExceptionReason;
exports.AccountAlreadyClosedException = AccountAlreadyClosedException;
exports.AccountAlreadyRegisteredException = AccountAlreadyRegisteredException;
exports.AccountFilterSensitiveLog = AccountFilterSensitiveLog;
exports.AccountJoinedMethod = AccountJoinedMethod;
exports.AccountNotFoundException = AccountNotFoundException;
exports.AccountNotRegisteredException = AccountNotRegisteredException;
exports.AccountOwnerNotVerifiedException = AccountOwnerNotVerifiedException;
exports.AccountState = AccountState;
exports.AccountStatus = AccountStatus;
exports.ActionType = ActionType;
exports.AlreadyInOrganizationException = AlreadyInOrganizationException;
exports.AttachPolicyCommand = AttachPolicyCommand;
exports.CancelHandshakeCommand = CancelHandshakeCommand;
exports.CancelHandshakeResponseFilterSensitiveLog = CancelHandshakeResponseFilterSensitiveLog;
exports.ChildNotFoundException = ChildNotFoundException;
exports.ChildType = ChildType;
exports.CloseAccountCommand = CloseAccountCommand;
exports.ConcurrentModificationException = ConcurrentModificationException;
exports.ConflictException = ConflictException;
exports.ConstraintViolationException = ConstraintViolationException;
exports.ConstraintViolationExceptionReason = ConstraintViolationExceptionReason;
exports.CreateAccountCommand = CreateAccountCommand;
exports.CreateAccountFailureReason = CreateAccountFailureReason;
exports.CreateAccountRequestFilterSensitiveLog = CreateAccountRequestFilterSensitiveLog;
exports.CreateAccountResponseFilterSensitiveLog = CreateAccountResponseFilterSensitiveLog;
exports.CreateAccountState = CreateAccountState;
exports.CreateAccountStatusFilterSensitiveLog = CreateAccountStatusFilterSensitiveLog;
exports.CreateAccountStatusNotFoundException = CreateAccountStatusNotFoundException;
exports.CreateGovCloudAccountCommand = CreateGovCloudAccountCommand;
exports.CreateGovCloudAccountRequestFilterSensitiveLog = CreateGovCloudAccountRequestFilterSensitiveLog;
exports.CreateGovCloudAccountResponseFilterSensitiveLog = CreateGovCloudAccountResponseFilterSensitiveLog;
exports.CreateOrganizationCommand = CreateOrganizationCommand;
exports.CreateOrganizationResponseFilterSensitiveLog = CreateOrganizationResponseFilterSensitiveLog;
exports.CreateOrganizationalUnitCommand = CreateOrganizationalUnitCommand;
exports.CreatePolicyCommand = CreatePolicyCommand;
exports.DeclineHandshakeCommand = DeclineHandshakeCommand;
exports.DeclineHandshakeResponseFilterSensitiveLog = DeclineHandshakeResponseFilterSensitiveLog;
exports.DelegatedAdministratorFilterSensitiveLog = DelegatedAdministratorFilterSensitiveLog;
exports.DeleteOrganizationCommand = DeleteOrganizationCommand;
exports.DeleteOrganizationalUnitCommand = DeleteOrganizationalUnitCommand;
exports.DeletePolicyCommand = DeletePolicyCommand;
exports.DeleteResourcePolicyCommand = DeleteResourcePolicyCommand;
exports.DeregisterDelegatedAdministratorCommand = DeregisterDelegatedAdministratorCommand;
exports.DescribeAccountCommand = DescribeAccountCommand;
exports.DescribeAccountResponseFilterSensitiveLog = DescribeAccountResponseFilterSensitiveLog;
exports.DescribeCreateAccountStatusCommand = DescribeCreateAccountStatusCommand;
exports.DescribeCreateAccountStatusResponseFilterSensitiveLog = DescribeCreateAccountStatusResponseFilterSensitiveLog;
exports.DescribeEffectivePolicyCommand = DescribeEffectivePolicyCommand;
exports.DescribeHandshakeCommand = DescribeHandshakeCommand;
exports.DescribeHandshakeResponseFilterSensitiveLog = DescribeHandshakeResponseFilterSensitiveLog;
exports.DescribeOrganizationCommand = DescribeOrganizationCommand;
exports.DescribeOrganizationResponseFilterSensitiveLog = DescribeOrganizationResponseFilterSensitiveLog;
exports.DescribeOrganizationalUnitCommand = DescribeOrganizationalUnitCommand;
exports.DescribePolicyCommand = DescribePolicyCommand;
exports.DescribeResourcePolicyCommand = DescribeResourcePolicyCommand;
exports.DestinationParentNotFoundException = DestinationParentNotFoundException;
exports.DetachPolicyCommand = DetachPolicyCommand;
exports.DisableAWSServiceAccessCommand = DisableAWSServiceAccessCommand;
exports.DisablePolicyTypeCommand = DisablePolicyTypeCommand;
exports.DuplicateAccountException = DuplicateAccountException;
exports.DuplicateHandshakeException = DuplicateHandshakeException;
exports.DuplicateOrganizationalUnitException = DuplicateOrganizationalUnitException;
exports.DuplicatePolicyAttachmentException = DuplicatePolicyAttachmentException;
exports.DuplicatePolicyException = DuplicatePolicyException;
exports.EffectivePolicyNotFoundException = EffectivePolicyNotFoundException;
exports.EffectivePolicyType = EffectivePolicyType;
exports.EnableAWSServiceAccessCommand = EnableAWSServiceAccessCommand;
exports.EnableAllFeaturesCommand = EnableAllFeaturesCommand;
exports.EnableAllFeaturesResponseFilterSensitiveLog = EnableAllFeaturesResponseFilterSensitiveLog;
exports.EnablePolicyTypeCommand = EnablePolicyTypeCommand;
exports.FinalizingOrganizationException = FinalizingOrganizationException;
exports.HandshakeAlreadyInStateException = HandshakeAlreadyInStateException;
exports.HandshakeConstraintViolationException = HandshakeConstraintViolationException;
exports.HandshakeConstraintViolationExceptionReason = HandshakeConstraintViolationExceptionReason;
exports.HandshakeFilterSensitiveLog = HandshakeFilterSensitiveLog;
exports.HandshakeNotFoundException = HandshakeNotFoundException;
exports.HandshakePartyFilterSensitiveLog = HandshakePartyFilterSensitiveLog;
exports.HandshakePartyType = HandshakePartyType;
exports.HandshakeResourceFilterSensitiveLog = HandshakeResourceFilterSensitiveLog;
exports.HandshakeResourceType = HandshakeResourceType;
exports.HandshakeState = HandshakeState;
exports.IAMUserAccessToBilling = IAMUserAccessToBilling;
exports.InvalidHandshakeTransitionException = InvalidHandshakeTransitionException;
exports.InvalidInputException = InvalidInputException;
exports.InvalidInputExceptionReason = InvalidInputExceptionReason;
exports.InviteAccountToOrganizationCommand = InviteAccountToOrganizationCommand;
exports.InviteAccountToOrganizationRequestFilterSensitiveLog = InviteAccountToOrganizationRequestFilterSensitiveLog;
exports.InviteAccountToOrganizationResponseFilterSensitiveLog = InviteAccountToOrganizationResponseFilterSensitiveLog;
exports.LeaveOrganizationCommand = LeaveOrganizationCommand;
exports.ListAWSServiceAccessForOrganizationCommand = ListAWSServiceAccessForOrganizationCommand;
exports.ListAccountsCommand = ListAccountsCommand;
exports.ListAccountsForParentCommand = ListAccountsForParentCommand;
exports.ListAccountsForParentResponseFilterSensitiveLog = ListAccountsForParentResponseFilterSensitiveLog;
exports.ListAccountsResponseFilterSensitiveLog = ListAccountsResponseFilterSensitiveLog;
exports.ListAccountsWithInvalidEffectivePolicyCommand = ListAccountsWithInvalidEffectivePolicyCommand;
exports.ListAccountsWithInvalidEffectivePolicyResponseFilterSensitiveLog = ListAccountsWithInvalidEffectivePolicyResponseFilterSensitiveLog;
exports.ListChildrenCommand = ListChildrenCommand;
exports.ListCreateAccountStatusCommand = ListCreateAccountStatusCommand;
exports.ListCreateAccountStatusResponseFilterSensitiveLog = ListCreateAccountStatusResponseFilterSensitiveLog;
exports.ListDelegatedAdministratorsCommand = ListDelegatedAdministratorsCommand;
exports.ListDelegatedAdministratorsResponseFilterSensitiveLog = ListDelegatedAdministratorsResponseFilterSensitiveLog;
exports.ListDelegatedServicesForAccountCommand = ListDelegatedServicesForAccountCommand;
exports.ListEffectivePolicyValidationErrorsCommand = ListEffectivePolicyValidationErrorsCommand;
exports.ListHandshakesForAccountCommand = ListHandshakesForAccountCommand;
exports.ListHandshakesForAccountResponseFilterSensitiveLog = ListHandshakesForAccountResponseFilterSensitiveLog;
exports.ListHandshakesForOrganizationCommand = ListHandshakesForOrganizationCommand;
exports.ListHandshakesForOrganizationResponseFilterSensitiveLog = ListHandshakesForOrganizationResponseFilterSensitiveLog;
exports.ListOrganizationalUnitsForParentCommand = ListOrganizationalUnitsForParentCommand;
exports.ListParentsCommand = ListParentsCommand;
exports.ListPoliciesCommand = ListPoliciesCommand;
exports.ListPoliciesForTargetCommand = ListPoliciesForTargetCommand;
exports.ListRootsCommand = ListRootsCommand;
exports.ListTagsForResourceCommand = ListTagsForResourceCommand;
exports.ListTargetsForPolicyCommand = ListTargetsForPolicyCommand;
exports.MalformedPolicyDocumentException = MalformedPolicyDocumentException;
exports.MasterCannotLeaveOrganizationException = MasterCannotLeaveOrganizationException;
exports.MoveAccountCommand = MoveAccountCommand;
exports.OrganizationFeatureSet = OrganizationFeatureSet;
exports.OrganizationFilterSensitiveLog = OrganizationFilterSensitiveLog;
exports.OrganizationNotEmptyException = OrganizationNotEmptyException;
exports.OrganizationalUnitNotEmptyException = OrganizationalUnitNotEmptyException;
exports.OrganizationalUnitNotFoundException = OrganizationalUnitNotFoundException;
exports.Organizations = Organizations;
exports.OrganizationsClient = OrganizationsClient;
exports.OrganizationsServiceException = OrganizationsServiceException;
exports.ParentNotFoundException = ParentNotFoundException;
exports.ParentType = ParentType;
exports.PolicyChangesInProgressException = PolicyChangesInProgressException;
exports.PolicyInUseException = PolicyInUseException;
exports.PolicyNotAttachedException = PolicyNotAttachedException;
exports.PolicyNotFoundException = PolicyNotFoundException;
exports.PolicyType = PolicyType;
exports.PolicyTypeAlreadyEnabledException = PolicyTypeAlreadyEnabledException;
exports.PolicyTypeNotAvailableForOrganizationException = PolicyTypeNotAvailableForOrganizationException;
exports.PolicyTypeNotEnabledException = PolicyTypeNotEnabledException;
exports.PolicyTypeStatus = PolicyTypeStatus;
exports.PutResourcePolicyCommand = PutResourcePolicyCommand;
exports.RegisterDelegatedAdministratorCommand = RegisterDelegatedAdministratorCommand;
exports.RemoveAccountFromOrganizationCommand = RemoveAccountFromOrganizationCommand;
exports.ResourcePolicyNotFoundException = ResourcePolicyNotFoundException;
exports.RootNotFoundException = RootNotFoundException;
exports.ServiceException = ServiceException;
exports.SourceParentNotFoundException = SourceParentNotFoundException;
exports.TagResourceCommand = TagResourceCommand;
exports.TargetNotFoundException = TargetNotFoundException;
exports.TargetType = TargetType;
exports.TooManyRequestsException = TooManyRequestsException;
exports.UnsupportedAPIEndpointException = UnsupportedAPIEndpointException;
exports.UntagResourceCommand = UntagResourceCommand;
exports.UpdateOrganizationalUnitCommand = UpdateOrganizationalUnitCommand;
exports.UpdatePolicyCommand = UpdatePolicyCommand;
exports.paginateListAWSServiceAccessForOrganization = paginateListAWSServiceAccessForOrganization;
exports.paginateListAccounts = paginateListAccounts;
exports.paginateListAccountsForParent = paginateListAccountsForParent;
exports.paginateListAccountsWithInvalidEffectivePolicy = paginateListAccountsWithInvalidEffectivePolicy;
exports.paginateListChildren = paginateListChildren;
exports.paginateListCreateAccountStatus = paginateListCreateAccountStatus;
exports.paginateListDelegatedAdministrators = paginateListDelegatedAdministrators;
exports.paginateListDelegatedServicesForAccount = paginateListDelegatedServicesForAccount;
exports.paginateListEffectivePolicyValidationErrors = paginateListEffectivePolicyValidationErrors;
exports.paginateListHandshakesForAccount = paginateListHandshakesForAccount;
exports.paginateListHandshakesForOrganization = paginateListHandshakesForOrganization;
exports.paginateListOrganizationalUnitsForParent = paginateListOrganizationalUnitsForParent;
exports.paginateListParents = paginateListParents;
exports.paginateListPolicies = paginateListPolicies;
exports.paginateListPoliciesForTarget = paginateListPoliciesForTarget;
exports.paginateListRoots = paginateListRoots;
exports.paginateListTagsForResource = paginateListTagsForResource;
exports.paginateListTargetsForPolicy = paginateListTargetsForPolicy;
