Index: dev/workflows/generate_versions.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/dev/workflows/generate_versions.py b/dev/workflows/generate_versions.py
new file mode 100644
--- /dev/null	(date 1762589336279)
+++ b/dev/workflows/generate_versions.py	(date 1762589336279)
@@ -0,0 +1,9 @@
+# from __future__ import annotations
+#
+#
+# def main() -> None:
+#     pass
+#
+#
+# if __name__ == "__main__":
+#     main()
