// @aws-sdk/client-controltower/dist-es/models/models_0.js
import { ControlTowerServiceException as __BaseException } from "./ControlTowerServiceException";
export class AccessDeniedException extends __BaseException {
    name = "AccessDeniedException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
    }
}
export const BaselineOperationType = {
    DISABLE_BASELINE: "DISABLE_BASELINE",
    ENABLE_BASELINE: "ENABLE_BASELINE",
    RESET_ENABLED_BASELINE: "RESET_ENABLED_BASELINE",
    UPDATE_ENABLED_BASELINE: "UPDATE_ENABLED_BASELINE",
};
export const BaselineOperationStatus = {
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
    SUCCEEDED: "SUCCEEDED",
};
export class InternalServerException extends __BaseException {
    name = "InternalServerException";
    $fault = "server";
    $retryable = {};
    constructor(opts) {
        super({
            name: "InternalServerException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalServerException.prototype);
    }
}
export class ResourceNotFoundException extends __BaseException {
    name = "ResourceNotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
    }
}
export class ThrottlingException extends __BaseException {
    name = "ThrottlingException";
    $fault = "client";
    $retryable = {
        throttling: true,
    };
    serviceCode;
    quotaCode;
    retryAfterSeconds;
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottlingException.prototype);
        this.serviceCode = opts.serviceCode;
        this.quotaCode = opts.quotaCode;
        this.retryAfterSeconds = opts.retryAfterSeconds;
    }
}
export class ValidationException extends __BaseException {
    name = "ValidationException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ValidationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ValidationException.prototype);
    }
}
export const ControlOperationType = {
    DISABLE_CONTROL: "DISABLE_CONTROL",
    ENABLE_CONTROL: "ENABLE_CONTROL",
    RESET_ENABLED_CONTROL: "RESET_ENABLED_CONTROL",
    UPDATE_ENABLED_CONTROL: "UPDATE_ENABLED_CONTROL",
};
export const ControlOperationStatus = {
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
    SUCCEEDED: "SUCCEEDED",
};
export class ConflictException extends __BaseException {
    name = "ConflictException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ConflictException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConflictException.prototype);
    }
}
export class ServiceQuotaExceededException extends __BaseException {
    name = "ServiceQuotaExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ServiceQuotaExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceQuotaExceededException.prototype);
    }
}
export const EnabledBaselineDriftStatus = {
    DRIFTED: "DRIFTED",
    IN_SYNC: "IN_SYNC",
};
export const EnablementStatus = {
    FAILED: "FAILED",
    SUCCEEDED: "SUCCEEDED",
    UNDER_CHANGE: "UNDER_CHANGE",
};
export const DriftStatus = {
    DRIFTED: "DRIFTED",
    IN_SYNC: "IN_SYNC",
    NOT_CHECKING: "NOT_CHECKING",
    UNKNOWN: "UNKNOWN",
};
export const LandingZoneOperationType = {
    CREATE: "CREATE",
    DELETE: "DELETE",
    RESET: "RESET",
    UPDATE: "UPDATE",
};
export const LandingZoneOperationStatus = {
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
    SUCCEEDED: "SUCCEEDED",
};
export const LandingZoneDriftStatus = {
    DRIFTED: "DRIFTED",
    IN_SYNC: "IN_SYNC",
};
export const LandingZoneStatus = {
    ACTIVE: "ACTIVE",
    FAILED: "FAILED",
    PROCESSING: "PROCESSING",
};
