// @aws-sdk/client-controltower/dist-cjs/index.js
'use strict';

var middlewareHostHeader = require('@aws-sdk/middleware-host-header');
var middlewareLogger = require('@aws-sdk/middleware-logger');
var middlewareRecursionDetection = require('@aws-sdk/middleware-recursion-detection');
var middlewareUserAgent = require('@aws-sdk/middleware-user-agent');
var configResolver = require('@smithy/config-resolver');
var core = require('@smithy/core');
var middlewareContentLength = require('@smithy/middleware-content-length');
var middlewareEndpoint = require('@smithy/middleware-endpoint');
var middlewareRetry = require('@smithy/middleware-retry');
var smithyClient = require('@smithy/smithy-client');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var regionConfigResolver = require('@aws-sdk/region-config-resolver');
var protocolHttp = require('@smithy/protocol-http');
var middlewareSerde = require('@smithy/middleware-serde');
var core$1 = require('@aws-sdk/core');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "controltower",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(regionConfigResolver.getAwsRegionExtensionConfiguration(runtimeConfig), smithyClient.getDefaultExtensionConfiguration(runtimeConfig), protocolHttp.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, regionConfigResolver.resolveAwsRegionExtensionConfiguration(extensionConfiguration), smithyClient.resolveDefaultRuntimeConfig(extensionConfiguration), protocolHttp.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class ControlTowerClient extends smithyClient.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = middlewareUserAgent.resolveUserAgentConfig(_config_1);
        const _config_3 = middlewareRetry.resolveRetryConfig(_config_2);
        const _config_4 = configResolver.resolveRegionConfig(_config_3);
        const _config_5 = middlewareHostHeader.resolveHostHeaderConfig(_config_4);
        const _config_6 = middlewareEndpoint.resolveEndpointConfig(_config_5);
        const _config_7 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(middlewareUserAgent.getUserAgentPlugin(this.config));
        this.middlewareStack.use(middlewareRetry.getRetryPlugin(this.config));
        this.middlewareStack.use(middlewareContentLength.getContentLengthPlugin(this.config));
        this.middlewareStack.use(middlewareHostHeader.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(middlewareLogger.getLoggerPlugin(this.config));
        this.middlewareStack.use(middlewareRecursionDetection.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultControlTowerHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class ControlTowerServiceException extends smithyClient.ServiceException {
    constructor(options) {
        super(options);
        Object.setPrototypeOf(this, ControlTowerServiceException.prototype);
    }
}

class AccessDeniedException extends ControlTowerServiceException {
    name = "AccessDeniedException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
    }
}
const BaselineOperationType = {
    DISABLE_BASELINE: "DISABLE_BASELINE",
    ENABLE_BASELINE: "ENABLE_BASELINE",
    RESET_ENABLED_BASELINE: "RESET_ENABLED_BASELINE",
    UPDATE_ENABLED_BASELINE: "UPDATE_ENABLED_BASELINE",
};
const BaselineOperationStatus = {
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
    SUCCEEDED: "SUCCEEDED",
};
class InternalServerException extends ControlTowerServiceException {
    name = "InternalServerException";
    $fault = "server";
    $retryable = {};
    constructor(opts) {
        super({
            name: "InternalServerException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalServerException.prototype);
    }
}
class ResourceNotFoundException extends ControlTowerServiceException {
    name = "ResourceNotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
    }
}
class ThrottlingException extends ControlTowerServiceException {
    name = "ThrottlingException";
    $fault = "client";
    $retryable = {
        throttling: true,
    };
    serviceCode;
    quotaCode;
    retryAfterSeconds;
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottlingException.prototype);
        this.serviceCode = opts.serviceCode;
        this.quotaCode = opts.quotaCode;
        this.retryAfterSeconds = opts.retryAfterSeconds;
    }
}
class ValidationException extends ControlTowerServiceException {
    name = "ValidationException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ValidationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ValidationException.prototype);
    }
}
const ControlOperationType = {
    DISABLE_CONTROL: "DISABLE_CONTROL",
    ENABLE_CONTROL: "ENABLE_CONTROL",
    RESET_ENABLED_CONTROL: "RESET_ENABLED_CONTROL",
    UPDATE_ENABLED_CONTROL: "UPDATE_ENABLED_CONTROL",
};
const ControlOperationStatus = {
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
    SUCCEEDED: "SUCCEEDED",
};
class ConflictException extends ControlTowerServiceException {
    name = "ConflictException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ConflictException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConflictException.prototype);
    }
}
class ServiceQuotaExceededException extends ControlTowerServiceException {
    name = "ServiceQuotaExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ServiceQuotaExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceQuotaExceededException.prototype);
    }
}
const EnabledBaselineDriftStatus = {
    DRIFTED: "DRIFTED",
    IN_SYNC: "IN_SYNC",
};
const EnablementStatus = {
    FAILED: "FAILED",
    SUCCEEDED: "SUCCEEDED",
    UNDER_CHANGE: "UNDER_CHANGE",
};
const DriftStatus = {
    DRIFTED: "DRIFTED",
    IN_SYNC: "IN_SYNC",
    NOT_CHECKING: "NOT_CHECKING",
    UNKNOWN: "UNKNOWN",
};
const LandingZoneOperationType = {
    CREATE: "CREATE",
    DELETE: "DELETE",
    RESET: "RESET",
    UPDATE: "UPDATE",
};
const LandingZoneOperationStatus = {
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
    SUCCEEDED: "SUCCEEDED",
};
const LandingZoneDriftStatus = {
    DRIFTED: "DRIFTED",
    IN_SYNC: "IN_SYNC",
};
const LandingZoneStatus = {
    ACTIVE: "ACTIVE",
    FAILED: "FAILED",
    PROCESSING: "PROCESSING",
};

const se_CreateLandingZoneCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/create-landingzone");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        manifest: (_) => se_Manifest(_),
        tags: (_) => smithyClient._json(_),
        version: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_DeleteLandingZoneCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/delete-landingzone");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        landingZoneIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_DisableBaselineCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/disable-baseline");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        enabledBaselineIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_DisableControlCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/disable-control");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        controlIdentifier: [],
        targetIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_EnableBaselineCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/enable-baseline");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        baselineIdentifier: [],
        baselineVersion: [],
        parameters: (_) => se_EnabledBaselineParameters(_),
        tags: (_) => smithyClient._json(_),
        targetIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_EnableControlCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/enable-control");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        controlIdentifier: [],
        parameters: (_) => se_EnabledControlParameters(_),
        tags: (_) => smithyClient._json(_),
        targetIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_GetBaselineCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/get-baseline");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        baselineIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_GetBaselineOperationCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/get-baseline-operation");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        operationIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_GetControlOperationCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/get-control-operation");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        operationIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_GetEnabledBaselineCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/get-enabled-baseline");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        enabledBaselineIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_GetEnabledControlCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/get-enabled-control");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        enabledControlIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_GetLandingZoneCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/get-landingzone");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        landingZoneIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_GetLandingZoneOperationCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/get-landingzone-operation");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        operationIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ListBaselinesCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/list-baselines");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        maxResults: [],
        nextToken: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ListControlOperationsCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/list-control-operations");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        filter: (_) => smithyClient._json(_),
        maxResults: [],
        nextToken: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ListEnabledBaselinesCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/list-enabled-baselines");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        filter: (_) => smithyClient._json(_),
        includeChildren: [],
        maxResults: [],
        nextToken: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ListEnabledControlsCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/list-enabled-controls");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        filter: (_) => smithyClient._json(_),
        maxResults: [],
        nextToken: [],
        targetIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ListLandingZoneOperationsCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/list-landingzone-operations");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        filter: (_) => smithyClient._json(_),
        maxResults: [],
        nextToken: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ListLandingZonesCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/list-landingzones");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        maxResults: [],
        nextToken: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ListTagsForResourceCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {};
    b.bp("/tags/{resourceArn}");
    b.p("resourceArn", () => input.resourceArn, "{resourceArn}", false);
    let body;
    b.m("GET").h(headers).b(body);
    return b.build();
};
const se_ResetEnabledBaselineCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/reset-enabled-baseline");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        enabledBaselineIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ResetEnabledControlCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/reset-enabled-control");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        enabledControlIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_ResetLandingZoneCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/reset-landingzone");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        landingZoneIdentifier: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_TagResourceCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/tags/{resourceArn}");
    b.p("resourceArn", () => input.resourceArn, "{resourceArn}", false);
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        tags: (_) => smithyClient._json(_),
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_UntagResourceCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {};
    b.bp("/tags/{resourceArn}");
    b.p("resourceArn", () => input.resourceArn, "{resourceArn}", false);
    const query = smithyClient.map({
        [_tK]: [smithyClient.expectNonNull(input.tagKeys, `tagKeys`) != null, () => input[_tK] || []],
    });
    let body;
    b.m("DELETE").h(headers).q(query).b(body);
    return b.build();
};
const se_UpdateEnabledBaselineCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/update-enabled-baseline");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        baselineVersion: [],
        enabledBaselineIdentifier: [],
        parameters: (_) => se_EnabledBaselineParameters(_),
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_UpdateEnabledControlCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/update-enabled-control");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        enabledControlIdentifier: [],
        parameters: (_) => se_EnabledControlParameters(_),
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const se_UpdateLandingZoneCommand = async (input, context) => {
    const b = core.requestBuilder(input, context);
    const headers = {
        "content-type": "application/json",
    };
    b.bp("/update-landingzone");
    let body;
    body = JSON.stringify(smithyClient.take(input, {
        landingZoneIdentifier: [],
        manifest: (_) => se_Manifest(_),
        version: [],
    }));
    b.m("POST").h(headers).b(body);
    return b.build();
};
const de_CreateLandingZoneCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        arn: smithyClient.expectString,
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_DeleteLandingZoneCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_DisableBaselineCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_DisableControlCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_EnableBaselineCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        arn: smithyClient.expectString,
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_EnableControlCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        arn: smithyClient.expectString,
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_GetBaselineCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        arn: smithyClient.expectString,
        description: smithyClient.expectString,
        name: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_GetBaselineOperationCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        baselineOperation: (_) => de_BaselineOperation(_),
    });
    Object.assign(contents, doc);
    return contents;
};
const de_GetControlOperationCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        controlOperation: (_) => de_ControlOperation(_),
    });
    Object.assign(contents, doc);
    return contents;
};
const de_GetEnabledBaselineCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        enabledBaselineDetails: (_) => de_EnabledBaselineDetails(_),
    });
    Object.assign(contents, doc);
    return contents;
};
const de_GetEnabledControlCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        enabledControlDetails: (_) => de_EnabledControlDetails(_),
    });
    Object.assign(contents, doc);
    return contents;
};
const de_GetLandingZoneCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        landingZone: (_) => de_LandingZoneDetail(_),
    });
    Object.assign(contents, doc);
    return contents;
};
const de_GetLandingZoneOperationCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationDetails: (_) => de_LandingZoneOperationDetail(_),
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ListBaselinesCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        baselines: smithyClient._json,
        nextToken: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ListControlOperationsCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        controlOperations: (_) => de_ControlOperations(_),
        nextToken: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ListEnabledBaselinesCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        enabledBaselines: smithyClient._json,
        nextToken: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ListEnabledControlsCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        enabledControls: smithyClient._json,
        nextToken: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ListLandingZoneOperationsCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        landingZoneOperations: smithyClient._json,
        nextToken: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ListLandingZonesCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        landingZones: smithyClient._json,
        nextToken: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ListTagsForResourceCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        tags: smithyClient._json,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ResetEnabledBaselineCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ResetEnabledControlCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_ResetLandingZoneCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_TagResourceCommand = async (output, context) => {
    if (output.statusCode !== 204 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    await smithyClient.collectBody(output.body, context);
    return contents;
};
const de_UntagResourceCommand = async (output, context) => {
    if (output.statusCode !== 204 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    await smithyClient.collectBody(output.body, context);
    return contents;
};
const de_UpdateEnabledBaselineCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_UpdateEnabledControlCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_UpdateLandingZoneCommand = async (output, context) => {
    if (output.statusCode !== 200 && output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const contents = smithyClient.map({
        $metadata: deserializeMetadata(output),
    });
    const data = smithyClient.expectNonNull(smithyClient.expectObject(await core$1.parseJsonBody(output.body, context)), "body");
    const doc = smithyClient.take(data, {
        operationIdentifier: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    return contents;
};
const de_CommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await core$1.parseJsonErrorBody(output.body, context),
    };
    const errorCode = core$1.loadRestJsonErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "AccessDeniedException":
        case "com.amazonaws.controltower#AccessDeniedException":
            throw await de_AccessDeniedExceptionRes(parsedOutput);
        case "ConflictException":
        case "com.amazonaws.controltower#ConflictException":
            throw await de_ConflictExceptionRes(parsedOutput);
        case "InternalServerException":
        case "com.amazonaws.controltower#InternalServerException":
            throw await de_InternalServerExceptionRes(parsedOutput);
        case "ThrottlingException":
        case "com.amazonaws.controltower#ThrottlingException":
            throw await de_ThrottlingExceptionRes(parsedOutput);
        case "ValidationException":
        case "com.amazonaws.controltower#ValidationException":
            throw await de_ValidationExceptionRes(parsedOutput);
        case "ResourceNotFoundException":
        case "com.amazonaws.controltower#ResourceNotFoundException":
            throw await de_ResourceNotFoundExceptionRes(parsedOutput);
        case "ServiceQuotaExceededException":
        case "com.amazonaws.controltower#ServiceQuotaExceededException":
            throw await de_ServiceQuotaExceededExceptionRes(parsedOutput);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody,
                errorCode,
            });
    }
};
const throwDefaultError = smithyClient.withBaseException(ControlTowerServiceException);
const de_AccessDeniedExceptionRes = async (parsedOutput, context) => {
    const contents = smithyClient.map({});
    const data = parsedOutput.body;
    const doc = smithyClient.take(data, {
        message: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    const exception = new AccessDeniedException({
        $metadata: deserializeMetadata(parsedOutput),
        ...contents,
    });
    return smithyClient.decorateServiceException(exception, parsedOutput.body);
};
const de_ConflictExceptionRes = async (parsedOutput, context) => {
    const contents = smithyClient.map({});
    const data = parsedOutput.body;
    const doc = smithyClient.take(data, {
        message: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    const exception = new ConflictException({
        $metadata: deserializeMetadata(parsedOutput),
        ...contents,
    });
    return smithyClient.decorateServiceException(exception, parsedOutput.body);
};
const de_InternalServerExceptionRes = async (parsedOutput, context) => {
    const contents = smithyClient.map({});
    const data = parsedOutput.body;
    const doc = smithyClient.take(data, {
        message: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    const exception = new InternalServerException({
        $metadata: deserializeMetadata(parsedOutput),
        ...contents,
    });
    return smithyClient.decorateServiceException(exception, parsedOutput.body);
};
const de_ResourceNotFoundExceptionRes = async (parsedOutput, context) => {
    const contents = smithyClient.map({});
    const data = parsedOutput.body;
    const doc = smithyClient.take(data, {
        message: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    const exception = new ResourceNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...contents,
    });
    return smithyClient.decorateServiceException(exception, parsedOutput.body);
};
const de_ServiceQuotaExceededExceptionRes = async (parsedOutput, context) => {
    const contents = smithyClient.map({});
    const data = parsedOutput.body;
    const doc = smithyClient.take(data, {
        message: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    const exception = new ServiceQuotaExceededException({
        $metadata: deserializeMetadata(parsedOutput),
        ...contents,
    });
    return smithyClient.decorateServiceException(exception, parsedOutput.body);
};
const de_ThrottlingExceptionRes = async (parsedOutput, context) => {
    const contents = smithyClient.map({
        [_rAS]: [() => void 0 !== parsedOutput.headers[_ra], () => smithyClient.strictParseInt32(parsedOutput.headers[_ra])],
    });
    const data = parsedOutput.body;
    const doc = smithyClient.take(data, {
        message: smithyClient.expectString,
        quotaCode: smithyClient.expectString,
        serviceCode: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    const exception = new ThrottlingException({
        $metadata: deserializeMetadata(parsedOutput),
        ...contents,
    });
    return smithyClient.decorateServiceException(exception, parsedOutput.body);
};
const de_ValidationExceptionRes = async (parsedOutput, context) => {
    const contents = smithyClient.map({});
    const data = parsedOutput.body;
    const doc = smithyClient.take(data, {
        message: smithyClient.expectString,
    });
    Object.assign(contents, doc);
    const exception = new ValidationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...contents,
    });
    return smithyClient.decorateServiceException(exception, parsedOutput.body);
};
const se_EnabledBaselineParameter = (input, context) => {
    return smithyClient.take(input, {
        key: [],
        value: (_) => se_EnabledBaselineParameterDocument(_),
    });
};
const se_EnabledBaselineParameterDocument = (input, context) => {
    return input;
};
const se_EnabledBaselineParameters = (input, context) => {
    return input
        .filter((e) => e != null)
        .map((entry) => {
        return se_EnabledBaselineParameter(entry);
    });
};
const se_EnabledControlParameter = (input, context) => {
    return smithyClient.take(input, {
        key: [],
        value: (_) => se_Document(_),
    });
};
const se_EnabledControlParameters = (input, context) => {
    return input
        .filter((e) => e != null)
        .map((entry) => {
        return se_EnabledControlParameter(entry);
    });
};
const se_Manifest = (input, context) => {
    return input;
};
const se_Document = (input, context) => {
    return input;
};
const de_BaselineOperation = (output, context) => {
    return smithyClient.take(output, {
        endTime: (_) => smithyClient.expectNonNull(smithyClient.parseRfc3339DateTimeWithOffset(_)),
        operationIdentifier: smithyClient.expectString,
        operationType: smithyClient.expectString,
        startTime: (_) => smithyClient.expectNonNull(smithyClient.parseRfc3339DateTimeWithOffset(_)),
        status: smithyClient.expectString,
        statusMessage: smithyClient.expectString,
    });
};
const de_ControlOperation = (output, context) => {
    return smithyClient.take(output, {
        controlIdentifier: smithyClient.expectString,
        enabledControlIdentifier: smithyClient.expectString,
        endTime: (_) => smithyClient.expectNonNull(smithyClient.parseRfc3339DateTimeWithOffset(_)),
        operationIdentifier: smithyClient.expectString,
        operationType: smithyClient.expectString,
        startTime: (_) => smithyClient.expectNonNull(smithyClient.parseRfc3339DateTimeWithOffset(_)),
        status: smithyClient.expectString,
        statusMessage: smithyClient.expectString,
        targetIdentifier: smithyClient.expectString,
    });
};
const de_ControlOperations = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_ControlOperationSummary(entry);
    });
    return retVal;
};
const de_ControlOperationSummary = (output, context) => {
    return smithyClient.take(output, {
        controlIdentifier: smithyClient.expectString,
        enabledControlIdentifier: smithyClient.expectString,
        endTime: (_) => smithyClient.expectNonNull(smithyClient.parseRfc3339DateTimeWithOffset(_)),
        operationIdentifier: smithyClient.expectString,
        operationType: smithyClient.expectString,
        startTime: (_) => smithyClient.expectNonNull(smithyClient.parseRfc3339DateTimeWithOffset(_)),
        status: smithyClient.expectString,
        statusMessage: smithyClient.expectString,
        targetIdentifier: smithyClient.expectString,
    });
};
const de_EnabledBaselineDetails = (output, context) => {
    return smithyClient.take(output, {
        arn: smithyClient.expectString,
        baselineIdentifier: smithyClient.expectString,
        baselineVersion: smithyClient.expectString,
        driftStatusSummary: smithyClient._json,
        parameters: (_) => de_EnabledBaselineParameterSummaries(_),
        parentIdentifier: smithyClient.expectString,
        statusSummary: smithyClient._json,
        targetIdentifier: smithyClient.expectString,
    });
};
const de_EnabledBaselineParameterDocument = (output, context) => {
    return output;
};
const de_EnabledBaselineParameterSummaries = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_EnabledBaselineParameterSummary(entry);
    });
    return retVal;
};
const de_EnabledBaselineParameterSummary = (output, context) => {
    return smithyClient.take(output, {
        key: smithyClient.expectString,
        value: (_) => de_EnabledBaselineParameterDocument(_),
    });
};
const de_EnabledControlDetails = (output, context) => {
    return smithyClient.take(output, {
        arn: smithyClient.expectString,
        controlIdentifier: smithyClient.expectString,
        driftStatusSummary: smithyClient._json,
        parameters: (_) => de_EnabledControlParameterSummaries(_),
        statusSummary: smithyClient._json,
        targetIdentifier: smithyClient.expectString,
        targetRegions: smithyClient._json,
    });
};
const de_EnabledControlParameterSummaries = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_EnabledControlParameterSummary(entry);
    });
    return retVal;
};
const de_EnabledControlParameterSummary = (output, context) => {
    return smithyClient.take(output, {
        key: smithyClient.expectString,
        value: (_) => de_Document(_),
    });
};
const de_LandingZoneDetail = (output, context) => {
    return smithyClient.take(output, {
        arn: smithyClient.expectString,
        driftStatus: smithyClient._json,
        latestAvailableVersion: smithyClient.expectString,
        manifest: (_) => de_Manifest(_),
        status: smithyClient.expectString,
        version: smithyClient.expectString,
    });
};
const de_LandingZoneOperationDetail = (output, context) => {
    return smithyClient.take(output, {
        endTime: (_) => smithyClient.expectNonNull(smithyClient.parseRfc3339DateTimeWithOffset(_)),
        operationIdentifier: smithyClient.expectString,
        operationType: smithyClient.expectString,
        startTime: (_) => smithyClient.expectNonNull(smithyClient.parseRfc3339DateTimeWithOffset(_)),
        status: smithyClient.expectString,
        statusMessage: smithyClient.expectString,
    });
};
const de_Manifest = (output, context) => {
    return output;
};
const de_Document = (output, context) => {
    return output;
};
const deserializeMetadata = (output) => ({
    httpStatusCode: output.statusCode,
    requestId: output.headers["x-amzn-requestid"] ?? output.headers["x-amzn-request-id"] ?? output.headers["x-amz-request-id"],
    extendedRequestId: output.headers["x-amz-id-2"],
    cfId: output.headers["x-amz-cf-id"],
});
const _rAS = "retryAfterSeconds";
const _ra = "retry-after";
const _tK = "tagKeys";

class CreateLandingZoneCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "CreateLandingZone", {})
    .n("ControlTowerClient", "CreateLandingZoneCommand")
    .f(void 0, void 0)
    .ser(se_CreateLandingZoneCommand)
    .de(de_CreateLandingZoneCommand)
    .build() {
}

class DeleteLandingZoneCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "DeleteLandingZone", {})
    .n("ControlTowerClient", "DeleteLandingZoneCommand")
    .f(void 0, void 0)
    .ser(se_DeleteLandingZoneCommand)
    .de(de_DeleteLandingZoneCommand)
    .build() {
}

class DisableBaselineCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "DisableBaseline", {})
    .n("ControlTowerClient", "DisableBaselineCommand")
    .f(void 0, void 0)
    .ser(se_DisableBaselineCommand)
    .de(de_DisableBaselineCommand)
    .build() {
}

class DisableControlCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "DisableControl", {})
    .n("ControlTowerClient", "DisableControlCommand")
    .f(void 0, void 0)
    .ser(se_DisableControlCommand)
    .de(de_DisableControlCommand)
    .build() {
}

class EnableBaselineCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "EnableBaseline", {})
    .n("ControlTowerClient", "EnableBaselineCommand")
    .f(void 0, void 0)
    .ser(se_EnableBaselineCommand)
    .de(de_EnableBaselineCommand)
    .build() {
}

class EnableControlCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "EnableControl", {})
    .n("ControlTowerClient", "EnableControlCommand")
    .f(void 0, void 0)
    .ser(se_EnableControlCommand)
    .de(de_EnableControlCommand)
    .build() {
}

class GetBaselineCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "GetBaseline", {})
    .n("ControlTowerClient", "GetBaselineCommand")
    .f(void 0, void 0)
    .ser(se_GetBaselineCommand)
    .de(de_GetBaselineCommand)
    .build() {
}

class GetBaselineOperationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "GetBaselineOperation", {})
    .n("ControlTowerClient", "GetBaselineOperationCommand")
    .f(void 0, void 0)
    .ser(se_GetBaselineOperationCommand)
    .de(de_GetBaselineOperationCommand)
    .build() {
}

class GetControlOperationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "GetControlOperation", {})
    .n("ControlTowerClient", "GetControlOperationCommand")
    .f(void 0, void 0)
    .ser(se_GetControlOperationCommand)
    .de(de_GetControlOperationCommand)
    .build() {
}

class GetEnabledBaselineCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "GetEnabledBaseline", {})
    .n("ControlTowerClient", "GetEnabledBaselineCommand")
    .f(void 0, void 0)
    .ser(se_GetEnabledBaselineCommand)
    .de(de_GetEnabledBaselineCommand)
    .build() {
}

class GetEnabledControlCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "GetEnabledControl", {})
    .n("ControlTowerClient", "GetEnabledControlCommand")
    .f(void 0, void 0)
    .ser(se_GetEnabledControlCommand)
    .de(de_GetEnabledControlCommand)
    .build() {
}

class GetLandingZoneCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "GetLandingZone", {})
    .n("ControlTowerClient", "GetLandingZoneCommand")
    .f(void 0, void 0)
    .ser(se_GetLandingZoneCommand)
    .de(de_GetLandingZoneCommand)
    .build() {
}

class GetLandingZoneOperationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "GetLandingZoneOperation", {})
    .n("ControlTowerClient", "GetLandingZoneOperationCommand")
    .f(void 0, void 0)
    .ser(se_GetLandingZoneOperationCommand)
    .de(de_GetLandingZoneOperationCommand)
    .build() {
}

class ListBaselinesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ListBaselines", {})
    .n("ControlTowerClient", "ListBaselinesCommand")
    .f(void 0, void 0)
    .ser(se_ListBaselinesCommand)
    .de(de_ListBaselinesCommand)
    .build() {
}

class ListControlOperationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ListControlOperations", {})
    .n("ControlTowerClient", "ListControlOperationsCommand")
    .f(void 0, void 0)
    .ser(se_ListControlOperationsCommand)
    .de(de_ListControlOperationsCommand)
    .build() {
}

class ListEnabledBaselinesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ListEnabledBaselines", {})
    .n("ControlTowerClient", "ListEnabledBaselinesCommand")
    .f(void 0, void 0)
    .ser(se_ListEnabledBaselinesCommand)
    .de(de_ListEnabledBaselinesCommand)
    .build() {
}

class ListEnabledControlsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ListEnabledControls", {})
    .n("ControlTowerClient", "ListEnabledControlsCommand")
    .f(void 0, void 0)
    .ser(se_ListEnabledControlsCommand)
    .de(de_ListEnabledControlsCommand)
    .build() {
}

class ListLandingZoneOperationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ListLandingZoneOperations", {})
    .n("ControlTowerClient", "ListLandingZoneOperationsCommand")
    .f(void 0, void 0)
    .ser(se_ListLandingZoneOperationsCommand)
    .de(de_ListLandingZoneOperationsCommand)
    .build() {
}

class ListLandingZonesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ListLandingZones", {})
    .n("ControlTowerClient", "ListLandingZonesCommand")
    .f(void 0, void 0)
    .ser(se_ListLandingZonesCommand)
    .de(de_ListLandingZonesCommand)
    .build() {
}

class ListTagsForResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ListTagsForResource", {})
    .n("ControlTowerClient", "ListTagsForResourceCommand")
    .f(void 0, void 0)
    .ser(se_ListTagsForResourceCommand)
    .de(de_ListTagsForResourceCommand)
    .build() {
}

class ResetEnabledBaselineCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ResetEnabledBaseline", {})
    .n("ControlTowerClient", "ResetEnabledBaselineCommand")
    .f(void 0, void 0)
    .ser(se_ResetEnabledBaselineCommand)
    .de(de_ResetEnabledBaselineCommand)
    .build() {
}

class ResetEnabledControlCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ResetEnabledControl", {})
    .n("ControlTowerClient", "ResetEnabledControlCommand")
    .f(void 0, void 0)
    .ser(se_ResetEnabledControlCommand)
    .de(de_ResetEnabledControlCommand)
    .build() {
}

class ResetLandingZoneCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "ResetLandingZone", {})
    .n("ControlTowerClient", "ResetLandingZoneCommand")
    .f(void 0, void 0)
    .ser(se_ResetLandingZoneCommand)
    .de(de_ResetLandingZoneCommand)
    .build() {
}

class TagResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "TagResource", {})
    .n("ControlTowerClient", "TagResourceCommand")
    .f(void 0, void 0)
    .ser(se_TagResourceCommand)
    .de(de_TagResourceCommand)
    .build() {
}

class UntagResourceCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "UntagResource", {})
    .n("ControlTowerClient", "UntagResourceCommand")
    .f(void 0, void 0)
    .ser(se_UntagResourceCommand)
    .de(de_UntagResourceCommand)
    .build() {
}

class UpdateEnabledBaselineCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "UpdateEnabledBaseline", {})
    .n("ControlTowerClient", "UpdateEnabledBaselineCommand")
    .f(void 0, void 0)
    .ser(se_UpdateEnabledBaselineCommand)
    .de(de_UpdateEnabledBaselineCommand)
    .build() {
}

class UpdateEnabledControlCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "UpdateEnabledControl", {})
    .n("ControlTowerClient", "UpdateEnabledControlCommand")
    .f(void 0, void 0)
    .ser(se_UpdateEnabledControlCommand)
    .de(de_UpdateEnabledControlCommand)
    .build() {
}

class UpdateLandingZoneCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("AWSControlTowerApis", "UpdateLandingZone", {})
    .n("ControlTowerClient", "UpdateLandingZoneCommand")
    .f(void 0, void 0)
    .ser(se_UpdateLandingZoneCommand)
    .de(de_UpdateLandingZoneCommand)
    .build() {
}

const commands = {
    CreateLandingZoneCommand,
    DeleteLandingZoneCommand,
    DisableBaselineCommand,
    DisableControlCommand,
    EnableBaselineCommand,
    EnableControlCommand,
    GetBaselineCommand,
    GetBaselineOperationCommand,
    GetControlOperationCommand,
    GetEnabledBaselineCommand,
    GetEnabledControlCommand,
    GetLandingZoneCommand,
    GetLandingZoneOperationCommand,
    ListBaselinesCommand,
    ListControlOperationsCommand,
    ListEnabledBaselinesCommand,
    ListEnabledControlsCommand,
    ListLandingZoneOperationsCommand,
    ListLandingZonesCommand,
    ListTagsForResourceCommand,
    ResetEnabledBaselineCommand,
    ResetEnabledControlCommand,
    ResetLandingZoneCommand,
    TagResourceCommand,
    UntagResourceCommand,
    UpdateEnabledBaselineCommand,
    UpdateEnabledControlCommand,
    UpdateLandingZoneCommand,
};
class ControlTower extends ControlTowerClient {
}
smithyClient.createAggregatedClient(commands, ControlTower);

const paginateListBaselines = core.createPaginator(ControlTowerClient, ListBaselinesCommand, "nextToken", "nextToken", "maxResults");

const paginateListControlOperations = core.createPaginator(ControlTowerClient, ListControlOperationsCommand, "nextToken", "nextToken", "maxResults");

const paginateListEnabledBaselines = core.createPaginator(ControlTowerClient, ListEnabledBaselinesCommand, "nextToken", "nextToken", "maxResults");

const paginateListEnabledControls = core.createPaginator(ControlTowerClient, ListEnabledControlsCommand, "nextToken", "nextToken", "maxResults");

const paginateListLandingZoneOperations = core.createPaginator(ControlTowerClient, ListLandingZoneOperationsCommand, "nextToken", "nextToken", "maxResults");

const paginateListLandingZones = core.createPaginator(ControlTowerClient, ListLandingZonesCommand, "nextToken", "nextToken", "maxResults");

Object.defineProperty(exports, "$Command", {
    enumerable: true,
    get: function () { return smithyClient.Command; }
});
Object.defineProperty(exports, "__Client", {
    enumerable: true,
    get: function () { return smithyClient.Client; }
});
exports.AccessDeniedException = AccessDeniedException;
exports.BaselineOperationStatus = BaselineOperationStatus;
exports.BaselineOperationType = BaselineOperationType;
exports.ConflictException = ConflictException;
exports.ControlOperationStatus = ControlOperationStatus;
exports.ControlOperationType = ControlOperationType;
exports.ControlTower = ControlTower;
exports.ControlTowerClient = ControlTowerClient;
exports.ControlTowerServiceException = ControlTowerServiceException;
exports.CreateLandingZoneCommand = CreateLandingZoneCommand;
exports.DeleteLandingZoneCommand = DeleteLandingZoneCommand;
exports.DisableBaselineCommand = DisableBaselineCommand;
exports.DisableControlCommand = DisableControlCommand;
exports.DriftStatus = DriftStatus;
exports.EnableBaselineCommand = EnableBaselineCommand;
exports.EnableControlCommand = EnableControlCommand;
exports.EnabledBaselineDriftStatus = EnabledBaselineDriftStatus;
exports.EnablementStatus = EnablementStatus;
exports.GetBaselineCommand = GetBaselineCommand;
exports.GetBaselineOperationCommand = GetBaselineOperationCommand;
exports.GetControlOperationCommand = GetControlOperationCommand;
exports.GetEnabledBaselineCommand = GetEnabledBaselineCommand;
exports.GetEnabledControlCommand = GetEnabledControlCommand;
exports.GetLandingZoneCommand = GetLandingZoneCommand;
exports.GetLandingZoneOperationCommand = GetLandingZoneOperationCommand;
exports.InternalServerException = InternalServerException;
exports.LandingZoneDriftStatus = LandingZoneDriftStatus;
exports.LandingZoneOperationStatus = LandingZoneOperationStatus;
exports.LandingZoneOperationType = LandingZoneOperationType;
exports.LandingZoneStatus = LandingZoneStatus;
exports.ListBaselinesCommand = ListBaselinesCommand;
exports.ListControlOperationsCommand = ListControlOperationsCommand;
exports.ListEnabledBaselinesCommand = ListEnabledBaselinesCommand;
exports.ListEnabledControlsCommand = ListEnabledControlsCommand;
exports.ListLandingZoneOperationsCommand = ListLandingZoneOperationsCommand;
exports.ListLandingZonesCommand = ListLandingZonesCommand;
exports.ListTagsForResourceCommand = ListTagsForResourceCommand;
exports.ResetEnabledBaselineCommand = ResetEnabledBaselineCommand;
exports.ResetEnabledControlCommand = ResetEnabledControlCommand;
exports.ResetLandingZoneCommand = ResetLandingZoneCommand;
exports.ResourceNotFoundException = ResourceNotFoundException;
exports.ServiceQuotaExceededException = ServiceQuotaExceededException;
exports.TagResourceCommand = TagResourceCommand;
exports.ThrottlingException = ThrottlingException;
exports.UntagResourceCommand = UntagResourceCommand;
exports.UpdateEnabledBaselineCommand = UpdateEnabledBaselineCommand;
exports.UpdateEnabledControlCommand = UpdateEnabledControlCommand;
exports.UpdateLandingZoneCommand = UpdateLandingZoneCommand;
exports.ValidationException = ValidationException;
exports.paginateListBaselines = paginateListBaselines;
exports.paginateListControlOperations = paginateListControlOperations;
exports.paginateListEnabledBaselines = paginateListEnabledBaselines;
exports.paginateListEnabledControls = paginateListEnabledControls;
exports.paginateListLandingZoneOperations = paginateListLandingZoneOperations;
exports.paginateListLandingZones = paginateListLandingZones;
